diff -ruNp ./src.vc10/libs/log/uqmlog.c ./src/libs/log/uqmlog.c
--- ./src.vc10/libs/log/uqmlog.c	2017-10-23 10:31:15 -0700
+++ ./src/libs/log/uqmlog.c	2017-10-23 10:40:22 -0700
@@ -185,7 +185,7 @@ void
 log_addV (log_Level level, const char *fmt, va_list list)
 {
 	log_Entry full_msg;
-	uqm_vsnprintf (full_msg, sizeof (full_msg) - 1, fmt, list);
+	vsnprintf (full_msg, sizeof (full_msg) - 1, fmt, list);
 	full_msg[sizeof (full_msg) - 1] = '\0';
 	
 	if ((int)level <= maxStreamLevel)
@@ -221,7 +221,7 @@ void
 log_add_nothreadV (log_Level level, const char *fmt, va_list list)
 {
 	log_Entry full_msg;
-	uqm_vsnprintf (full_msg, sizeof (full_msg) - 1, fmt, list);
+	vsnprintf (full_msg, sizeof (full_msg) - 1, fmt, list);
 	full_msg[sizeof (full_msg) - 1] = '\0';
 	
 	if ((int)level <= maxStreamLevel)
diff -ruNp ./src.vc10/libs/uio/utils.c ./src/libs/uio/utils.c
--- ./src.vc10/libs/uio/utils.c	2017-10-23 10:31:17 -0700
+++ ./src/libs/uio/utils.c	2017-10-23 10:40:22 -0700
@@ -391,7 +391,7 @@ snprintf(char *str, size_t size, const c
 // MSVC does not have vsnprintf(). It does have a _vsnprintf(), but it does
 // not \0-terminate a truncated string as the C standard prescribes.
 static inline int
-uqm_vsnprintf(char *str, size_t size, const char *format, va_list args)
+vsnprintf(char *str, size_t size, const char *format, va_list args)
 {
 	int result = _vsnprintf (str, size, format, args);
 	if (str != NULL && size != 0)
@@ -425,7 +425,7 @@ uio_vasprintf(const char *format, va_lis
 	}
 
 	for (;;) {
-		int printResult = uqm_vsnprintf(buf, bufSize, format, args);
+		int printResult = vsnprintf(buf, bufSize, format, args);
 		if (printResult < 0) {
 			// This means the buffer was not large enough, but vsnprintf()
 			// does not give us any clue on how large it should be.
diff -ruNp ./src.vc10/port.c ./src/port.c
--- ./src.vc10/port.c	2017-10-23 10:31:17 -0700
+++ ./src/port.c	2017-10-23 10:40:22 -0700
@@ -134,7 +134,7 @@ snprintf(char *str, size_t size, const c
 }
 
 int
-uqm_vsnprintf(char *str, size_t size, const char *format, va_list args)
+vsnprintf(char *str, size_t size, const char *format, va_list args)
 {
 	int result = _vsnprintf (str, size, format, args);
 	if (str != NULL && size != 0)
diff -ruNp ./src.vc10/port.h ./src/port.h
--- ./src.vc10/port.h	2017-10-23 10:31:17 -0700
+++ ./src/port.h	2017-10-23 10:40:22 -0700
@@ -177,7 +177,7 @@ typedef unsigned short mode_t;
 extern "C" {
 #endif
 int snprintf(char *str, size_t size, const char *format, ...);
-int uqm_vsnprintf(char *str, size_t size, const char *format, va_list args);
+int vsnprintf(char *str, size_t size, const char *format, va_list args);
 #if defined(__cplusplus)
 }
 #endif
diff -ruNp ./src.vc10/uqm.c ./src/uqm.c
--- ./src.vc10/uqm.c	2017-10-23 10:31:19 -0700
+++ ./src/uqm.c	2017-10-23 10:40:24 -0700
@@ -543,7 +543,7 @@ saveErrorV (const char *fmt, va_list lis
 		++len;
 		--left;
 	}
-	uqm_vsnprintf (errBuffer + len, left, fmt, list);
+	vsnprintf (errBuffer + len, left, fmt, list);
 	errBuffer[sizeof (errBuffer) - 1] = '\0';
 }
 
