diff -ruNp ./src.orig/libs/network/wspiapiwrap.h ./src/libs/network/wspiapiwrap.h
--- ./src.orig/libs/network/wspiapiwrap.h	2017-04-11 02:08:46 -0500
+++ ./src/libs/network/wspiapiwrap.h	2017-04-11 02:08:36 -0500
@@ -19,7 +19,10 @@
 #ifndef _WSPIAPIWRAP_H
 #define _WSPIAPIWRAP_H
 
-// HACK. See wspiapiwrap.c
+/* ************
+   DC: This whole section commented out due to VSC compiling issues. Found here: http://forum.uqm.stack.nl/index.php?topic=4353.0
+
+   // HACK. See wspiapiwrap.c
 #	define getaddrinfo WspiapiGetAddrInfo
 #	define getnameinfo WspiapiGetNameInfo
 #	define freeaddrinfo WspiapiFreeAddrInfo
@@ -28,6 +31,8 @@ int WINAPI WspiapiGetAddrInfo(const char
 		const struct addrinfo *hints, struct addrinfo **res);
 int WINAPI WspiapiGetNameInfo (const struct sockaddr *sa, socklen_t salen,
 		char *host, size_t hostlen, char *serv, size_t servlen, int flags);
+*/
+#include <wspiapi.h>  //DC: replaced upper section with this part to (hopefully) compile.
 
 #endif  /* _WSPIAPIWRAP_H */
 
diff -ruNp ./src.orig/libs/uio/utils.c ./src/libs/uio/utils.c
--- ./src.orig/libs/uio/utils.c	2017-04-11 02:08:48 -0500
+++ ./src/libs/uio/utils.c	2017-04-11 02:08:37 -0500
@@ -391,7 +391,7 @@ snprintf(char *str, size_t size, const c
 // MSVC does not have vsnprintf(). It does have a _vsnprintf(), but it does
 // not \0-terminate a truncated string as the C standard prescribes.
 static inline int
-vsnprintf(char *str, size_t size, const char *format, va_list args)
+uqm_vsnprintf(char *str, size_t size, const char *format, va_list args)
 {
 	int result = _vsnprintf (str, size, format, args);
 	if (str != NULL && size != 0)
@@ -425,7 +425,7 @@ uio_vasprintf(const char *format, va_lis
 	}
 
 	for (;;) {
-		int printResult = vsnprintf(buf, bufSize, format, args);
+		int printResult = uqm_vsnprintf(buf, bufSize, format, args);
 		if (printResult < 0) {
 			// This means the buffer was not large enough, but vsnprintf()
 			// does not give us any clue on how large it should be.
diff -ruNp ./src.orig/options.c ./src/options.c
--- ./src.orig/options.c	2017-04-11 02:08:48 -0500
+++ ./src/options.c	2017-04-11 02:08:37 -0500
@@ -52,6 +52,20 @@ int optSmoothScroll;
 int optMeleeScale;
 const char **optAddons;
 
+// Serosis
+BOOLEAN optCheatMode;
+BOOLEAN optGodMode;
+BOOLEAN optTimeDilation;
+BOOLEAN optBubbleWarp;
+BOOLEAN optRoseBud;
+BOOLEAN optUnlockShips;
+BOOLEAN optHeadStart;
+BOOLEAN optUnlockUpgrades;
+BOOLEAN optLanderMods;
+BOOLEAN optFastForward;
+BOOLEAN optSkipIntro;
+BOOLEAN optFMV;
+
 BOOLEAN opt3doMusic;
 BOOLEAN optRemixMusic;
 BOOLEAN optSubtitles;
@@ -358,6 +372,7 @@ mountAddonDir (uio_Repository *repositor
 	static uio_AutoMount *autoMount[] = { NULL };
 	uio_MountHandle *mountHandle;
 	uio_DirList *availableAddons;
+	char mountname[128]; // MB to JMS: no need to manually terminate string, snprintf does it for you. Moving mountname outside of loop.
 
 	if (addonDirName != NULL)
 	{
@@ -416,7 +431,6 @@ mountAddonDir (uio_Repository *repositor
 		count = 0;
 		for (i = 0; i < availableAddons->numNames; ++i)
 		{
-			static char mountname[128];
 			uio_DirHandle *addonDir;
 			const char *addon = availableAddons->names[i];
 			
@@ -426,8 +440,7 @@ mountAddonDir (uio_Repository *repositor
 			++count;
 			log_add (log_Info, "    %d. %s", count, addon);
 		
-			snprintf(mountname, 128, "addons/%s", addon);
-			mountname[127]=0;
+			snprintf (mountname, sizeof mountname, "addons/%s", addon);
 
 			addonDir = uio_openDirRelative (addonsDir, addon, 0);
 			if (addonDir == NULL)
diff -ruNp ./src.orig/options.h ./src/options.h
--- ./src.orig/options.h	2017-04-11 02:08:48 -0500
+++ ./src/options.h	2017-04-11 02:08:37 -0500
@@ -38,6 +38,20 @@ extern int optWhichShield;
 extern int optSmoothScroll;
 extern int optMeleeScale;
 
+// Serosis
+extern BOOLEAN optCheatMode;
+extern BOOLEAN optGodMode;
+extern BOOLEAN optTimeDilation;
+extern BOOLEAN optBubbleWarp;
+extern BOOLEAN optRoseBud;
+extern BOOLEAN optUnlockShips;
+extern BOOLEAN optHeadStart;
+extern BOOLEAN optUnlockUpgrades;
+extern BOOLEAN optLanderMods;
+extern BOOLEAN optFastForward;
+extern BOOLEAN optSkipIntro;
+extern BOOLEAN optFMV;
+
 extern BOOLEAN opt3doMusic;
 extern BOOLEAN optRemixMusic;
 extern BOOLEAN optSubtitles;
diff -ruNp ./src.orig/port.c ./src/port.c
--- ./src.orig/port.c	2017-04-11 02:08:48 -0500
+++ ./src/port.c	2017-04-11 02:08:37 -0500
@@ -115,8 +115,8 @@ readdir_r(DIR *dirp, struct dirent *entr
 #endif
 
 #ifdef _MSC_VER
-// MSVC does not have snprintf() and vsnprintf(). It does have a _snprintf()
-// and _vsnprintf(), but these do not terminate a truncated string as
+// MSVC does not have snprintf() and uqm_vsnprintf(). It does have a _snprintf()
+// and _uqm_vsnprintf(), but these do not terminate a truncated string as
 // the C standard prescribes.
 int
 snprintf(char *str, size_t size, const char *format, ...)
@@ -134,7 +134,7 @@ snprintf(char *str, size_t size, const c
 }
 
 int
-vsnprintf(char *str, size_t size, const char *format, va_list args)
+uqm_vsnprintf(char *str, size_t size, const char *format, va_list args)
 {
 	int result = _vsnprintf (str, size, format, args);
 	if (str != NULL && size != 0)
diff -ruNp ./src.orig/port.h ./src/port.h
--- ./src.orig/port.h	2017-04-11 02:08:48 -0500
+++ ./src/port.h	2017-04-11 02:08:37 -0500
@@ -162,7 +162,7 @@ typedef unsigned short mode_t;
 #	include <stdarg.h>
 // Defined in port.c
 int snprintf(char *str, size_t size, const char *format, ...);
-int vsnprintf(char *str, size_t size, const char *format, va_list args);
+int uqm_vsnprintf(char *str, size_t size, const char *format, va_list args);
 #endif  /* _MSC_VER */
 
 // setenv()
diff -ruNp ./src.orig/regex/regex_internal.h ./src/regex/regex_internal.h
--- ./src.orig/regex/regex_internal.h	2017-04-11 02:08:48 -0500
+++ ./src/regex/regex_internal.h	2017-04-11 02:08:37 -0500
@@ -80,7 +80,11 @@
 # define BE(expr, val) __builtin_expect (expr, val)
 #else
 # define BE(expr, val) (expr)
-# define inline
+#ifdef _MSC_VER
+	#define inline __inline
+#else
+	#define inline
+#endif
 #endif
 
 /* Number of bits in a byte.  */
diff -ruNp ./src.orig/res/UrQuanMasters.rc ./src/res/UrQuanMasters.rc
--- ./src.orig/res/UrQuanMasters.rc	2017-04-11 02:08:48 -0500
+++ ./src/res/UrQuanMasters.rc	2017-04-11 02:08:38 -0500
@@ -67,10 +67,10 @@ END
 
 // Icon with lowest ID value placed first to ensure application icon
 // remains consistent on all systems.
-SDL_app            ICON    PRELOAD DISCARDABLE "ur-quan-icon-alpha.ico"
+SDL_app            ICON    PRELOAD DISCARDABLE "kohr-ah1.ico"
 102                ICON    DISCARDABLE     "ur-quan-icon-std.ico"
 103                ICON    DISCARDABLE     "ur-quan1.ico"
-104                ICON    DISCARDABLE     "sis1.ico"
+104                ICON    DISCARDABLE     "ur-quan-icon-alpha.ico"
 105                ICON    DISCARDABLE     "ur-quan2.ico"
-106                ICON    DISCARDABLE     "kohr-ah1.ico"
+106                ICON    DISCARDABLE     "sis1.ico"
 107                ICON    DISCARDABLE     "starcon2.ico"
diff -ruNp ./src.orig/res/darwin/Info.plist ./src/res/darwin/Info.plist
--- ./src.orig/res/darwin/Info.plist	2017-04-11 02:08:48 -0500
+++ ./src/res/darwin/Info.plist	2017-04-11 02:08:37 -0500
@@ -1,11 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!DOCTYPE plist PUBLIC "-//Apple Computer//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
 	<key>CFBundleDevelopmentRegion</key>
 	<string>English</string>
 	<key>CFBundleExecutable</key>
-	<string>The Ur-Quan Masters</string>
+	<string>The Ur-Quan Masters MegaMod</string>
 	<key>CFBundleIconFile</key>
 	<string>The Ur-Quan Masters.icns</string>
 	<key>CFBundleInfoDictionaryVersion</key>
Files ./src.orig/res/darwin/The Ur-Quan Masters.icns and ./src/res/darwin/The Ur-Quan Masters.icns differ
diff -ruNp ./src.orig/uqm/comm/arilou/arilouc.c ./src/uqm/comm/arilou/arilouc.c
--- ./src.orig/uqm/comm/arilou/arilouc.c	2017-04-11 02:08:49 -0500
+++ ./src/uqm/comm/arilou/arilouc.c	2017-04-11 02:08:38 -0500
@@ -490,18 +490,20 @@ AngryHomeArilou (RESPONSE_REF R)
 		DISABLE_PHRASE (i_just_like_to_leave);
 	}
 
-	if (PHRASE_ENABLED (invaders_from_mars))
+	if (PHRASE_ENABLED (invaders_from_mars)){
 		Response (invaders_from_mars, AngryHomeArilou);
-	else
-	{
+	} else {
 		Response (bug_eyed_fruitcakes, ExitConversation);
 	}
-	if (PHRASE_ENABLED (why_should_i_trust))
+	if (PHRASE_ENABLED (why_should_i_trust)){
 		Response (why_should_i_trust, AngryHomeArilou);
-	else if (PHRASE_ENABLED (what_about_interference))
+	} else if (PHRASE_ENABLED (what_about_interference)){
 		Response (what_about_interference, AngryHomeArilou);
+	}
 	Response (ok_lets_be_friends, ArilouHome);
-	Response (i_just_like_to_leave, AngryHomeArilou);
+	if (PHRASE_ENABLED (i_just_like_to_leave)){
+		Response (i_just_like_to_leave, AngryHomeArilou);
+	}
 }
 
 static void
diff -ruNp ./src.orig/uqm/comm/melnorm/melnorm.c ./src/uqm/comm/melnorm/melnorm.c
--- ./src.orig/uqm/comm/melnorm/melnorm.c	2017-04-11 02:08:49 -0500
+++ ./src/uqm/comm/melnorm/melnorm.c	2017-04-11 02:08:38 -0500
@@ -32,7 +32,7 @@
 #include "uqm/planets/planets.h"
 		// for xxx_DISASTER
 #include "uqm/sis.h"
-
+#include "options.h"
 
 #define NUM_HISTORY_ITEMS 9
 #define NUM_EVENT_ITEMS 8
@@ -252,7 +252,10 @@ StripShip (COUNT fuel_required)
 {
 	BYTE i, which_module;
 	SBYTE crew_pods;
-
+	COUNT StorageBayCapacity = STORAGE_BAY_CAPACITY;
+	if(optLanderMods){
+		StorageBayCapacity = StorageBayCapacity <<= 1;
+	}
 	SET_GAME_STATE (MELNORME_RESCUE_REFUSED, 0);
 
 	crew_pods = -(SBYTE)(
@@ -322,10 +325,7 @@ StripShip (COUNT fuel_required)
 		if (fuel_required > capacity)
 			fuel_required = capacity;
 
-		bays = -(SBYTE)(
-				(SIS_copy.TotalElementMass + STORAGE_BAY_CAPACITY - 1)
-				/ STORAGE_BAY_CAPACITY
-				);
+		bays = -(SBYTE)( (SIS_copy.TotalElementMass + StorageBayCapacity - 1) / StorageBayCapacity );
 		for (i = 0; i < NUM_MODULE_SLOTS; ++i)
 		{
 			which_module = SIS_copy.ModuleSlots[i];
@@ -1321,6 +1321,10 @@ NatureOfConversation (RESPONSE_REF R)
 				break;
 		}
 		SET_GAME_STATE (MELNORME_YACK_STACK2, stack + 5);
+		// Tying in the purple question with a Credit/R.U. cheat
+		if (optRoseBud){
+			SET_GAME_STATE (WHY_MELNORME_PURPLE, 0); // This is to make sure that the purple question shows up.
+		} // Serosis
 	}
 
 	rainbow_mask = MAKE_WORD (
@@ -1360,7 +1364,19 @@ NatureOfConversation (RESPONSE_REF R)
 		}
 		else if (PLAYER_SAID (R, why_turned_purple))
 		{
-			SET_GAME_STATE (WHY_MELNORME_PURPLE, 1);
+			if (optRoseBud){
+				SET_GAME_STATE (WHY_MELNORME_PURPLE, 0); // Setting this to 0 makes sure the question is repeatable
+				DeltaCredit(1000); // Adds 1000 Credits
+				GLOBAL_SIS (ResUnits) += 1000; // Adds 1000 R.U.
+				// Make sure the RU/credit amount is redrawn:
+				if (LOBYTE (GLOBAL (CurrentActivity)) == IN_HYPERSPACE || LOBYTE (GLOBAL (CurrentActivity)) == IN_INTERPLANETARY) {
+					LockMutex (GraphicsLock);
+					DrawStatusMessage (NULL);
+					UnlockMutex (GraphicsLock);
+				}
+			} else {
+				SET_GAME_STATE (WHY_MELNORME_PURPLE, 1);
+			} // Serosis
 
 			NPCPhrase (TURNED_PURPLE_BECAUSE);
 		}
diff -ruNp ./src.orig/uqm/comm/rebel/rebel.c ./src/uqm/comm/rebel/rebel.c
--- ./src.orig/uqm/comm/rebel/rebel.c	2017-04-11 02:08:49 -0500
+++ ./src/uqm/comm/rebel/rebel.c	2017-04-11 02:08:38 -0500
@@ -46,19 +46,17 @@ static LOCDATA yehat_desc =
 		{ /* right hand-wing tapping keyboard; front guy */
 			4, /* StartIndex */
 			3, /* NumFrames */
-			YOYO_ANIM
-					| WAIT_TALKING, /* AnimFlags */
+			YOYO_ANIM | WAIT_TALKING, /* AnimFlags */
 			ONE_SECOND / 10, ONE_SECOND / 10, /* FrameRate */
-			ONE_SECOND, ONE_SECOND * 3, /* RestartRate */
+			ONE_SECOND / 4, ONE_SECOND / 2,/* RestartRate */
 			(1 << 6) | (1 << 7),
 		},
 		{ /* left hand-wing tapping keyboard; front guy */
 			7, /* StartIndex */
 			3, /* NumFrames */
-			YOYO_ANIM
-					| WAIT_TALKING, /* AnimFlags */
+			YOYO_ANIM | WAIT_TALKING, /* AnimFlags */
 			ONE_SECOND / 10, ONE_SECOND / 10, /* FrameRate */
-			ONE_SECOND, ONE_SECOND * 3, /* RestartRate */
+			ONE_SECOND / 4, ONE_SECOND / 2,/* RestartRate */
 			(1 << 6) | (1 << 7),
 		},
 		{
@@ -82,7 +80,7 @@ static LOCDATA yehat_desc =
 			5, /* NumFrames */
 			YOYO_ANIM, /* AnimFlags */
 			ONE_SECOND / 15, ONE_SECOND / 15, /* FrameRate */
-			ONE_SECOND * 10, ONE_SECOND * 3,/* RestartRate */
+			ONE_SECOND * 6, ONE_SECOND * 3,/* RestartRate */
 			(1 << 2) | (1 << 14),
 		},
 		{
@@ -90,25 +88,23 @@ static LOCDATA yehat_desc =
 			5, /* NumFrames */
 			YOYO_ANIM, /* AnimFlags */
 			ONE_SECOND / 15, ONE_SECOND / 15, /* FrameRate */
-			ONE_SECOND * 10, ONE_SECOND * 3,/* RestartRate */
+			ONE_SECOND * 6, ONE_SECOND * 3,/* RestartRate */
 			(1 << 3),
 		},
 		{ /* right arm-wing rising; front guy */
 			26, /* StartIndex */
 			2, /* NumFrames */
-			YOYO_ANIM
-					| WAIT_TALKING, /* AnimFlags */
+			YOYO_ANIM | WAIT_TALKING, /* AnimFlags */
 			ONE_SECOND / 15, ONE_SECOND / 15, /* FrameRate */
-			ONE_SECOND * 10, ONE_SECOND * 3,/* RestartRate */
+			ONE_SECOND * 6, ONE_SECOND * 3,/* RestartRate */
 			(1 << 0) | (1 << 1),
 		},
 		{ /* left arm-wing rising; front guy */
 			28, /* StartIndex */
 			2, /* NumFrames */
-			YOYO_ANIM
-					| WAIT_TALKING, /* AnimFlags */
+			YOYO_ANIM | WAIT_TALKING, /* AnimFlags */
 			ONE_SECOND / 15, ONE_SECOND / 15, /* FrameRate */
-			ONE_SECOND * 10, ONE_SECOND * 3,/* RestartRate */
+			ONE_SECOND * 6, ONE_SECOND * 3,/* RestartRate */
 			(1 << 0) | (1 << 1),
 		},
 		{
@@ -162,8 +158,7 @@ static LOCDATA yehat_desc =
 		{
 			48, /* StartIndex */
 			4, /* NumFrames */
-			YOYO_ANIM
-					| WAIT_TALKING, /* AnimFlags */
+			YOYO_ANIM | WAIT_TALKING, /* AnimFlags */
 			ONE_SECOND / 30, 0, /* FrameRate */
 			ONE_SECOND, ONE_SECOND * 3, /* RestartRate */
 			(1 << 2) | (1 << 4),
diff -ruNp ./src.orig/uqm/comm/starbas/starbas.c ./src/uqm/comm/starbas/starbas.c
--- ./src.orig/uqm/comm/starbas/starbas.c	2017-04-11 02:08:50 -0500
+++ ./src/uqm/comm/starbas/starbas.c	2017-04-11 02:08:39 -0500
@@ -303,6 +303,7 @@ HierarchyInfo (RESPONSE_REF R)
 #define HIERARCHY_ANDROSYNTH (1 << 3)
 #define HIERARCHY_ILWRATH (1 << 4)
 #define HIERARCHY_VUX (1 << 5)
+#define HIERARCHY_URQUAN (1 << 6)
 	static BYTE HierarchyMask = 0;
 
 	if (PLAYER_SAID (R, what_about_hierarchy))
@@ -339,8 +340,12 @@ HierarchyInfo (RESPONSE_REF R)
 	{
 		NPCPhrase (ABOUT_VUX);
 		HierarchyMask |= HIERARCHY_VUX;
-	}
-
+	}else if (PLAYER_SAID (R, urquan)) {
+		NPCPhrase (ABOUT_URQUAN);
+		HierarchyMask |= HIERARCHY_URQUAN;
+ 	}
+	if (!(HierarchyMask & HIERARCHY_URQUAN))
+		Response (urquan, HierarchyInfo);
 	if (!(HierarchyMask & HIERARCHY_MYCON))
 		Response (mycon, HierarchyInfo);
 	if (!(HierarchyMask & HIERARCHY_SPATHI))
diff -ruNp ./src.orig/uqm/comm/syreen/syreenc.c ./src/uqm/comm/syreen/syreenc.c
--- ./src.orig/uqm/comm/syreen/syreenc.c	2017-04-11 02:08:50 -0500
+++ ./src/uqm/comm/syreen/syreenc.c	2017-04-11 02:08:39 -0500
@@ -652,13 +652,12 @@ InitialSyreen (RESPONSE_REF R)
 	else if (PLAYER_SAID (R, we_are_vindicator0))
 	{
 		NPCPhrase (WELCOME_VINDICATOR0);
-		if (speechVolumeScale == 0.0f)
-		{
+		if (speechVolumeScale == 0.0f){
 			NPCPhrase (GLOBAL_PLAYER_NAME);
 			NPCPhrase (WELCOME_VINDICATOR1);
 			NPCPhrase (GLOBAL_SHIP_NAME);
-			NPCPhrase (WELCOME_VINDICATOR2);
 		}
+		NPCPhrase (WELCOME_VINDICATOR2);
 		NPCPhrase (HOW_CAN_YOU_BE_HERE);
 	}
 	else if (PLAYER_SAID (R, we_are_impressed))
diff -ruNp ./src.orig/uqm/comm/thradd/thraddc.c ./src/uqm/comm/thradd/thraddc.c
--- ./src.orig/uqm/comm/thradd/thraddc.c	2017-04-11 02:08:50 -0500
+++ ./src/uqm/comm/thradd/thraddc.c	2017-04-11 02:08:39 -0500
@@ -695,6 +695,8 @@ static void
 Intro (void)
 {
 	BYTE NumVisits;
+	HFLEETINFO hThradd = GetStarShipFromIndex (&GLOBAL (avail_race_q), THRADDASH_SHIP);
+	FLEET_INFO *ThraddPtr = LockFleetInfo (&GLOBAL (avail_race_q), hThradd);
 
 	if (LOBYTE (GLOBAL (CurrentActivity)) == WON_LAST_BATTLE)
 	{
@@ -704,7 +706,7 @@ Intro (void)
 		return;
 	}
 
-	if (GET_GAME_STATE (AQUA_HELIX))
+	if (GET_GAME_STATE (AQUA_HELIX) && ThraddPtr->allied_state != GOOD_GUY)
 	{
 		NumVisits = GET_GAME_STATE (HELIX_VISITS);
 		switch (NumVisits++)
diff -ruNp ./src.orig/uqm/comm/umgah/strings.h ./src/uqm/comm/umgah/strings.h
--- ./src.orig/uqm/comm/umgah/strings.h	2017-04-11 02:08:50 -0500
+++ ./src/uqm/comm/umgah/strings.h	2017-04-11 02:08:39 -0500
@@ -109,6 +109,7 @@ enum
 	arilou_told_us,
 	bye_zombie,
 	GOODBYE_ZOMBIE,
+	OUT_TAKES,
 };
 
 #endif /* _STRINGS_H */
diff -ruNp ./src.orig/uqm/comm/umgah/umgahc.c ./src/uqm/comm/umgah/umgahc.c
--- ./src.orig/uqm/comm/umgah/umgahc.c	2017-04-11 02:08:50 -0500
+++ ./src/uqm/comm/umgah/umgahc.c	2017-04-11 02:08:39 -0500
@@ -520,8 +520,11 @@ Intro (void)
 {
 	BYTE NumVisits;
 
-	if (GET_GAME_STATE (UMGAH_HOSTILE))
-	{
+	if (LOBYTE (GLOBAL (CurrentActivity)) == WON_LAST_BATTLE) {
+		NPCPhrase (OUT_TAKES);
+		SET_GAME_STATE (BATTLE_SEGUE, 0);
+		return;
+	} else if (GET_GAME_STATE (UMGAH_HOSTILE)) {
 		NumVisits = GET_GAME_STATE (UMGAH_VISITS);
 		switch (NumVisits++)
 		{
diff -ruNp ./src.orig/uqm/credits.c ./src/uqm/credits.c
--- ./src.orig/uqm/credits.c	2017-04-11 02:08:50 -0500
+++ ./src/uqm/credits.c	2017-04-11 02:08:39 -0500
@@ -647,7 +647,7 @@ FreeCredits (void)
 static void
 OutTakes (void)
 {
-#define NUM_OUTTAKES 15
+#define NUM_OUTTAKES 16
 	static CONVERSATION outtake_list[NUM_OUTTAKES] =
 	{
 		ZOQFOTPIK_CONVERSATION,
@@ -664,6 +664,7 @@ OutTakes (void)
 		URQUAN_CONVERSATION,
 		VUX_CONVERSATION,
 		BLACKURQ_CONVERSATION,
+		UMGAH_CONVERSATION,
 		ARILOU_CONVERSATION
 	};
 
diff -ruNp ./src.orig/uqm/fmv.c ./src/uqm/fmv.c
--- ./src.orig/uqm/fmv.c	2017-04-11 02:08:51 -0500
+++ ./src/uqm/fmv.c	2017-04-11 02:08:39 -0500
@@ -132,5 +132,17 @@ Victory (void)
 	FadeScreen (FadeAllToBlack, 0);
 }
 
+void
+Logo (void)
+{
+	ShowPresentation (LOGOPRES_STRTAB);
+	SleepThreadUntil (FadeScreen (FadeAllToBlack, ONE_SECOND / 2));
+}
 
+void
+Drumall (void)
+{
+	ShowPresentation (DRUMALLPRES_STRTAB);
+	SleepThreadUntil (FadeScreen (FadeAllToBlack, ONE_SECOND / 2));
+}
 
diff -ruNp ./src.orig/uqm/fmv.h ./src/uqm/fmv.h
--- ./src.orig/uqm/fmv.h	2017-04-11 02:08:51 -0500
+++ ./src/uqm/fmv.h	2017-04-11 02:08:39 -0500
@@ -26,6 +26,8 @@ extern void SplashScreen (void (* DoProc
 extern void Introduction (void);
 extern void Victory (void);
 extern void DoShipSpin (COUNT index, MUSIC_REF hMusic);
+extern void Logo (void);
+extern void Drumall (void);
 
 extern BOOLEAN ShowPresentation (RESOURCE presentation);
 
diff -ruNp ./src.orig/uqm/gameev.c ./src/uqm/gameev.c
--- ./src.orig/uqm/gameev.c	2017-04-11 02:08:51 -0500
+++ ./src/uqm/gameev.c	2017-04-11 02:08:39 -0500
@@ -25,7 +25,7 @@
 #include "hyper.h"
 #include "libs/compiler.h"
 #include "libs/mathlib.h"
-
+#include "options.h"
 
 static void arilou_entrance_event (void);
 static void arilou_exit_event (void);
@@ -40,6 +40,8 @@ static void mycon_mission (void);
 
 void
 AddInitialGameEvents (void) {
+	COUNT kohrah_winning_years = optCheatMode ? YEARS_TO_KOHRAH_VICTORY + 25 : YEARS_TO_KOHRAH_VICTORY;
+
 	AddEvent (RELATIVE_EVENT, 0, 1, 0, HYPERSPACE_ENCOUNTER_EVENT);
 	AddEvent (ABSOLUTE_EVENT, 3, 17, START_YEAR, ARILOU_ENTRANCE_EVENT);
 	AddEvent (RELATIVE_EVENT, 0, 0, YEARS_TO_KOHRAH_VICTORY,
@@ -146,6 +148,7 @@ EventHandler (BYTE selector)
 				if (SpathiPtr->actual_strength)
 				{
 					ActivateStarShip (SPATHI_SHIP, SET_NOT_ALLIED);
+					ActivateStarShip (SPATHI_SHIP, REMOVE_BUILT);
 					SET_GAME_STATE (SPATHI_SHIELDED_SELVES, 1);
 					SpathiPtr->actual_strength = 0;
 				}
@@ -388,6 +391,9 @@ black_urquan_genocide (void)
 		else if (speed > 255)
 			speed = 255;
 
+		if (optCheatMode)
+			speed = 0;
+
 		SET_GAME_STATE (KOHR_AH_FRENZY, 1);
 		SET_GAME_STATE (KOHR_AH_VISITS, 0);
 		SET_GAME_STATE (KOHR_AH_REASONS, 0);
@@ -526,76 +532,63 @@ thradd_mission (void)
 static void
 ilwrath_mission (void)
 {
-	BYTE ThraddState;
-	HFLEETINFO hIlwrath, hThradd;
-	FLEET_INFO *IlwrathPtr;
-	FLEET_INFO *ThraddPtr;
+	COUNT MaddLength = 128;
+	SIZE strength_loss;
+	BYTE ThraddState = GET_GAME_STATE (THRADD_MISSION);
+	HFLEETINFO	hIlwrath = GetStarShipFromIndex (&GLOBAL (avail_race_q), ILWRATH_SHIP),
+				hThradd = GetStarShipFromIndex (&GLOBAL (avail_race_q), THRADDASH_SHIP);
+	FLEET_INFO	*IlwrathPtr = LockFleetInfo (&GLOBAL (avail_race_q), hIlwrath), 
+				*ThraddPtr = LockFleetInfo (&GLOBAL (avail_race_q), hThradd);
 
-	hIlwrath = GetStarShipFromIndex (&GLOBAL (avail_race_q), ILWRATH_SHIP);
-	IlwrathPtr = LockFleetInfo (&GLOBAL (avail_race_q), hIlwrath);
-	hThradd = GetStarShipFromIndex (&GLOBAL (avail_race_q), THRADDASH_SHIP);
-	ThraddPtr = LockFleetInfo (&GLOBAL (avail_race_q), hThradd);
-
-	if (IlwrathPtr->loc.x == ((2500 + 2535) >> 1)
-			&& IlwrathPtr->loc.y == ((8070 + 8358) >> 1))
-	{
+	if (IlwrathPtr->loc.x == ((2500 + 2535) >> 1) && IlwrathPtr->loc.y == ((8070 + 8358) >> 1)) { // What is with the silly math?
 		IlwrathPtr->actual_strength = 0;
-		ThraddPtr->actual_strength = 0;
 		IlwrathPtr->allied_state = DEAD_GUY;
-		ThraddPtr->allied_state = DEAD_GUY;
-	}
-	else if (IlwrathPtr->actual_strength)
-	{
-		if (!GET_GAME_STATE (ILWRATH_FIGHT_THRADDASH)
-				&& (IlwrathPtr->dest_loc.x != 2500
-				|| IlwrathPtr->dest_loc.y != 8070))
-		{
-			SetRaceDest (ILWRATH_SHIP, 2500, 8070, 90,
-					ADVANCE_ILWRATH_MISSION);
-		}
-		else
-		{
-#define MADD_LENGTH 128
-			SIZE strength_loss;
-
-			if (IlwrathPtr->days_left == 0)
-			{	/* arrived for battle */
+		if(ThraddPtr->allied_state != GOOD_GUY){
+			ThraddPtr->actual_strength = 0;
+			ThraddPtr->allied_state = DEAD_GUY;
+		}
+	} else if (IlwrathPtr->actual_strength) {
+		if (!GET_GAME_STATE (ILWRATH_FIGHT_THRADDASH) && (IlwrathPtr->dest_loc.x != 2500 || IlwrathPtr->dest_loc.y != 8070)) {
+			SetRaceDest (ILWRATH_SHIP, 2500, 8070, 90, ADVANCE_ILWRATH_MISSION); //90
+		} else {
+			if (IlwrathPtr->days_left == 0) {	/* arrived for battle */
 				SET_GAME_STATE (ILWRATH_FIGHT_THRADDASH, 1);
 				SET_GAME_STATE (HELIX_UNPROTECTED, 1);
 				strength_loss = (SIZE)IlwrathPtr->actual_strength;
-				IlwrathPtr->growth = (BYTE)(-strength_loss / MADD_LENGTH);
-				IlwrathPtr->growth_fract =
-						(BYTE)(((strength_loss % MADD_LENGTH) << 8) / MADD_LENGTH);
-				SetRaceDest (ILWRATH_SHIP,
-						(2500 + 2535) >> 1, (8070 + 8358) >> 1,
-						MADD_LENGTH - 1, ADVANCE_ILWRATH_MISSION);
-
-				strength_loss = (SIZE)ThraddPtr->actual_strength;
-				ThraddPtr->growth = (BYTE)(-strength_loss / MADD_LENGTH);
-				ThraddPtr->growth_fract =
-						(BYTE)(((strength_loss % MADD_LENGTH) << 8) / MADD_LENGTH);
-
-				SET_GAME_STATE (THRADD_VISITS, 0);
-				if (ThraddPtr->allied_state == GOOD_GUY)
-					ActivateStarShip (THRADDASH_SHIP, SET_NOT_ALLIED);
-			}
-
-			ThraddState = GET_GAME_STATE (THRADD_MISSION);
-			if (ThraddState == 0 || ThraddState > 3)
-			{	/* never went to Kohr-Ah or returned */
-				SetRaceDest (THRADDASH_SHIP,
-						(2500 + 2535) >> 1, (8070 + 8358) >> 1,
-						IlwrathPtr->days_left + 1, (BYTE)~0);
-			}
-			else if (ThraddState < 3)
-			{	/* recall on the double */
-				SetRaceDest (THRADDASH_SHIP, 2535, 8358, 10,
-						ADVANCE_THRADD_MISSION);
+				IlwrathPtr->growth = (BYTE)(-strength_loss / MaddLength);
+				IlwrathPtr->growth_fract = (BYTE)(((strength_loss % MaddLength) << 8) / MaddLength);
+				SetRaceDest (ILWRATH_SHIP, 2517, 8214, MaddLength - 1, ADVANCE_ILWRATH_MISSION);
+
+				if (ThraddPtr->allied_state == GOOD_GUY){
+					strength_loss = (SIZE)(ThraddPtr->actual_strength * 0.25); // Smarterer math
+					ThraddPtr->growth = (BYTE)(-strength_loss / MaddLength);
+					ThraddPtr->growth_fract = (BYTE)(((strength_loss % MaddLength) << 8) / MaddLength);
+					ThraddPtr->growth_err_term = 255 >> 1;
+				} else {
+					SET_GAME_STATE (THRADD_VISITS, 0);
+					strength_loss = (SIZE)ThraddPtr->actual_strength;
+					ThraddPtr->growth = (BYTE)(-strength_loss / MaddLength);
+					ThraddPtr->growth_fract = (BYTE)(((strength_loss % MaddLength) << 8) / MaddLength);
+				}
+			}
+
+			if (ThraddState == 0 || ThraddState > 3) {	/* never went to Kohr-Ah or returned */
+				SetRaceDest (THRADDASH_SHIP, (2500 + 2535) >> 1, (8070 + 8358) >> 1, IlwrathPtr->days_left + 1, (BYTE)~0);
+			} else if (ThraddState < 3) {	/* recall on the double */
+				SetRaceDest (THRADDASH_SHIP, 2535, 8358, 10, ADVANCE_THRADD_MISSION);
 				SET_GAME_STATE (THRADD_MISSION, 3);
 			}
 		}
 	}
-
+	if(ThraddPtr->allied_state == GOOD_GUY && !IlwrathPtr->actual_strength){		
+		ThraddPtr->growth = 0;
+		ThraddPtr->growth_fract = 0;
+		SET_GAME_STATE (ILWRATH_FIGHT_THRADDASH, 0);		
+		SetRaceDest (THRADDASH_SHIP, 2535, 8358, 3, (BYTE)~0);
+		if(!GET_GAME_STATE(AQUA_HELIX)){
+			SET_GAME_STATE (HELIX_UNPROTECTED, 0);
+		}
+	}
 	UnlockFleetInfo (&GLOBAL (avail_race_q), hThradd);
 	UnlockFleetInfo (&GLOBAL (avail_race_q), hIlwrath);
 }
diff -ruNp ./src.orig/uqm/gameinp.c ./src/uqm/gameinp.c
--- ./src.orig/uqm/gameinp.c	2017-04-11 02:08:51 -0500
+++ ./src/uqm/gameinp.c	2017-04-11 02:08:39 -0500
@@ -31,10 +31,10 @@
 #include "libs/inplib.h"
 #include "libs/timelib.h"
 #include "libs/threadlib.h"
+#include "setup.h"
 
-
-#define ACCELERATION_INCREMENT (ONE_SECOND / 12)
-#define MENU_REPEAT_DELAY (ONE_SECOND / 2)
+#define ACCELERATION_INCREMENT (ONE_SECOND / 28)
+#define MENU_REPEAT_DELAY (ONE_SECOND / 3)
 
 
 typedef struct
@@ -432,10 +432,18 @@ ControlInputToBattleInput (const int *ke
 		InputState |= BATTLE_LEFT;
 	if (keyState[KEY_RIGHT])
 		InputState |= BATTLE_RIGHT;
-	if (keyState[KEY_WEAPON])
+	if (keyState[KEY_WEAPON]){
+		if (optGodMode && PlayerControl[1] & COMPUTER_CONTROL){
+			resetEnergyBattle();
+		}
 		InputState |= BATTLE_WEAPON;
-	if (keyState[KEY_SPECIAL])
+	}
+	if (keyState[KEY_SPECIAL]){
+		if (optGodMode && PlayerControl[1] & COMPUTER_CONTROL){
+			resetEnergyBattle();
+		}
 		InputState |= BATTLE_SPECIAL;
+	}
 	if (keyState[KEY_ESCAPE])
 		InputState |= BATTLE_ESCAPE;
 	if (keyState[KEY_DOWN])
diff -ruNp ./src.orig/uqm/gameopt.c ./src/uqm/gameopt.c
--- ./src.orig/uqm/gameopt.c	2017-04-11 02:08:51 -0500
+++ ./src/uqm/gameopt.c	2017-04-11 02:08:39 -0500
@@ -40,7 +40,7 @@
 
 extern FRAME PlayFrame;
 
-#define MAX_SAVED_GAMES 50
+#define MAX_SAVED_GAMES 100
 #define SUMMARY_X_OFFS 14
 #define SUMMARY_SIDE_OFFS 7
 #define SAVES_PER_PAGE 5
diff -ruNp ./src.orig/uqm/globdata.c ./src/uqm/globdata.c
--- ./src.orig/uqm/globdata.c	2017-04-11 02:08:51 -0500
+++ ./src/uqm/globdata.c	2017-04-11 02:08:40 -0500
@@ -30,12 +30,12 @@
 #include "state.h"
 #include "grpinfo.h"
 #include "gamestr.h"
-
+#include "options.h"
 #include <stdlib.h>
 #ifdef STATE_DEBUG
 #	include "libs/log.h"
 #endif
-
+#include "uqmdebug.h"
 
 static void CreateRadar (void);
 
@@ -285,26 +285,59 @@ InitGameStructures (void)
 	GLOBAL (ElementWorth[PRECIOUS]) = 6;
 	GLOBAL_SIS (ElementAmounts[PRECIOUS]) = 0;
 	GLOBAL (ElementWorth[RADIOACTIVE]) = 8;
-	GLOBAL_SIS (ElementAmounts[RADIOACTIVE]) = 0;
+	if (optHeadStart){
+		GLOBAL_SIS (ElementAmounts[RADIOACTIVE]) = 1000;
+		GLOBAL_SIS (TotalElementMass) = 1000;
+		GLOBAL_SIS (TotalBioMass) += 1000;
+	} else {
+		GLOBAL_SIS (ElementAmounts[RADIOACTIVE]) = 0;
+	}
 	GLOBAL (ElementWorth[EXOTIC]) = 25;
 	GLOBAL_SIS (ElementAmounts[EXOTIC]) = 0;
 
-	for (i = 0; i < NUM_DRIVE_SLOTS; ++i)
-		GLOBAL_SIS (DriveSlots[i]) = EMPTY_SLOT + 0;
-	GLOBAL_SIS (DriveSlots[5]) =
-			GLOBAL_SIS (DriveSlots[6]) = FUSION_THRUSTER;
-	for (i = 0; i < NUM_JET_SLOTS; ++i)
-		GLOBAL_SIS (JetSlots[i]) = EMPTY_SLOT + 1;
-	GLOBAL_SIS (JetSlots[0]) =
-			GLOBAL_SIS (JetSlots[6]) = TURNING_JETS;
-	for (i = 0; i < NUM_MODULE_SLOTS; ++i)
-		GLOBAL_SIS (ModuleSlots[i]) = EMPTY_SLOT + 2;
-	GLOBAL_SIS (ModuleSlots[15]) = GUN_WEAPON;
-	GLOBAL_SIS (ModuleSlots[2]) = CREW_POD;
-	GLOBAL_SIS (CrewEnlisted) = CREW_POD_CAPACITY;
-	GLOBAL_SIS (ModuleSlots[8]) = STORAGE_BAY;
-	GLOBAL_SIS (ModuleSlots[1]) = FUEL_TANK;
-	GLOBAL_SIS (FuelOnBoard) = 10 * FUEL_TANK_SCALE;
+	if (optHeadStart){		
+		for (i = 0; i < NUM_DRIVE_SLOTS; i++)
+			GLOBAL_SIS (DriveSlots[i]) = FUSION_THRUSTER;
+		for (i = 0; i < NUM_JET_SLOTS; i++)
+			GLOBAL_SIS (JetSlots[i]) = TURNING_JETS;
+		for (i = 0; i < NUM_MODULE_SLOTS; ++i)
+			GLOBAL_SIS (ModuleSlots[i]) = EMPTY_SLOT + 2;
+		GLOBAL_SIS (ModuleSlots[0]) = FUEL_TANK;
+		GLOBAL_SIS (ModuleSlots[1]) = FUEL_TANK;
+		GLOBAL_SIS (ModuleSlots[2]) = CREW_POD;
+		GLOBAL_SIS (ModuleSlots[3]) = CREW_POD;
+		GLOBAL_SIS (ModuleSlots[4]) = STORAGE_BAY;
+		GLOBAL_SIS (ModuleSlots[5]) = STORAGE_BAY;
+		//GLOBAL_SIS (ModuleSlots[6]);
+		//GLOBAL_SIS (ModuleSlots[7]);
+		//GLOBAL_SIS (ModuleSlots[8]);
+		//GLOBAL_SIS (ModuleSlots[9]);
+		GLOBAL_SIS (ModuleSlots[10]) = DYNAMO_UNIT;
+		GLOBAL_SIS (ModuleSlots[11]) = DYNAMO_UNIT;
+		GLOBAL_SIS (ModuleSlots[12]) = DYNAMO_UNIT;
+		GLOBAL_SIS (ModuleSlots[13]) = DYNAMO_UNIT;
+		GLOBAL_SIS (ModuleSlots[14]) = GUN_WEAPON;
+		GLOBAL_SIS (ModuleSlots[15]) = GUN_WEAPON;
+		GLOBAL_SIS (FuelOnBoard) = 110 * FUEL_TANK_SCALE;
+		GLOBAL_SIS (CrewEnlisted) = 100;
+	} else {
+		for (i = 0; i < NUM_DRIVE_SLOTS; ++i)
+			GLOBAL_SIS (DriveSlots[i]) = EMPTY_SLOT + 0;
+		GLOBAL_SIS (DriveSlots[5]) =
+				GLOBAL_SIS (DriveSlots[6]) = FUSION_THRUSTER;
+		for (i = 0; i < NUM_JET_SLOTS; ++i)
+			GLOBAL_SIS (JetSlots[i]) = EMPTY_SLOT + 1;
+		GLOBAL_SIS (JetSlots[0]) =
+				GLOBAL_SIS (JetSlots[6]) = TURNING_JETS;
+		for (i = 0; i < NUM_MODULE_SLOTS; ++i)
+			GLOBAL_SIS (ModuleSlots[i]) = EMPTY_SLOT + 2;
+		GLOBAL_SIS (ModuleSlots[15]) = GUN_WEAPON;
+		GLOBAL_SIS (ModuleSlots[2]) = CREW_POD;
+		GLOBAL_SIS (CrewEnlisted) = CREW_POD_CAPACITY;
+		GLOBAL_SIS (ModuleSlots[8]) = STORAGE_BAY;
+		GLOBAL_SIS (ModuleSlots[1]) = FUEL_TANK;
+		GLOBAL_SIS (FuelOnBoard) = 10 * FUEL_TANK_SCALE;
+	}
 
 	InitQueue (&GLOBAL (built_ship_q),
 			MAX_BUILT_SHIPS, sizeof (SHIP_FRAGMENT));
@@ -328,7 +361,11 @@ InitGameStructures (void)
 	GLOBAL (ModuleCost[DYNAMO_UNIT]) = 2000 / MODULE_COST_SCALE;
 	GLOBAL (ModuleCost[GUN_WEAPON]) = 2000 / MODULE_COST_SCALE;
 
-	GLOBAL_SIS (NumLanders) = 1;
+	if (optHeadStart){
+		GLOBAL_SIS (NumLanders) = MAX_LANDERS;
+	} else {
+		GLOBAL_SIS (NumLanders) = 1;
+	}
 
 	utf8StringCopy (GLOBAL_SIS (ShipName), sizeof (GLOBAL_SIS (ShipName)),
 			GAME_STRING (NAMING_STRING_BASE + 2));
diff -ruNp ./src.orig/uqm/istrtab.h ./src/uqm/istrtab.h
--- ./src.orig/uqm/istrtab.h	2017-04-11 02:08:51 -0500
+++ ./src/uqm/istrtab.h	2017-04-11 02:08:40 -0500
@@ -37,6 +37,7 @@
 #define CYANIC_XLAT_TAB "planet.cyanic.translatetable"
 #define CYA_GAS_COLOR_TAB "planet.cyangas.colortable"
 #define CYA_GAS_XLAT_TAB "planet.cyangas.translatetable"
+#define DRUMALLPRES_STRTAB "slides.drumall"
 #define DRUUGE_RUINS_STRTAB "text.sphere"
 #define DUST_COLOR_TAB "planet.dust.colortable"
 #define DUST_XLAT_TAB "planet.dust.translatetable"
@@ -67,6 +68,7 @@
 #define JOYSTICK_ALPHA_STRTAB "text.joyalpha"
 #define LANTHANIDE_COLOR_TAB "planet.lanthanide.colortable"
 #define LANTHANIDE_XLAT_TAB "planet.lanthanide.translatetable"
+#define LOGOPRES_STRTAB "slides.logo"
 #define MAGMA_COLOR_TAB "planet.magma.colortable"
 #define MAGMA_XLAT_TAB "planet.magma.translatetable"
 #define MAGNETIC_COLOR_TAB "planet.magnetic.colortable"
diff -ruNp ./src.orig/uqm/misc.c ./src/uqm/misc.c
--- ./src.orig/uqm/misc.c	2017-04-11 02:08:51 -0500
+++ ./src/uqm/misc.c	2017-04-11 02:08:40 -0500
@@ -25,7 +25,9 @@
 #include "sounds.h"
 #include "weapon.h"
 #include "libs/mathlib.h"
-
+#include "options.h" // For God Mode invincibility
+#include "settings.h"
+#include "intel.h"
 
 void
 spawn_planet (void)
@@ -194,6 +196,10 @@ spawn_asteroid (ELEMENT *ElementPtr)
 void
 do_damage (ELEMENT *ElementPtr, SIZE damage)
 {
+	// God Mode, borrowed from the UQM-HD debug invincibility code
+	if (optGodMode && PlayerControl[1] & COMPUTER_CONTROL && ElementPtr->playerNr == 0){
+		damage = 0;
+	}
 	if (ElementPtr->state_flags & PLAYER_SHIP)
 	{
 		if (!DeltaCrew (ElementPtr, -damage))
diff -ruNp ./src.orig/uqm/outfit.c ./src/uqm/outfit.c
--- ./src.orig/uqm/outfit.c	2017-04-11 02:08:51 -0500
+++ ./src/uqm/outfit.c	2017-04-11 02:08:40 -0500
@@ -182,6 +182,10 @@ DoInstallModule (MENU_STATE *pMS)
 	BYTE NewState, new_slot_piece, old_slot_piece;
 	SIZE FirstItem, LastItem;
 	BOOLEAN select, cancel, motion;
+	COUNT StorageBayCapacity = STORAGE_BAY_CAPACITY;
+	if(optLanderMods){
+		StorageBayCapacity = StorageBayCapacity <<= 1;
+	}
 
 	if (GLOBAL (CurrentActivity) & CHECK_ABORT)
 	{
@@ -290,7 +294,7 @@ DoInstallModule (MENU_STATE *pMS)
 				}
 				else if (old_slot_piece == STORAGE_BAY)
 				{
-					if (GLOBAL_SIS (TotalElementMass) > STORAGE_BAY_CAPACITY
+					if (GLOBAL_SIS (TotalElementMass) > StorageBayCapacity
 							* (CountSISPieces (STORAGE_BAY) - 1))
 					{	// storage bay still needed for the cargo
 						PlayMenuSound (MENU_SOUND_FAILURE);
diff -ruNp ./src.orig/uqm/planets/generate/genthrad.c ./src/uqm/planets/generate/genthrad.c
--- ./src.orig/uqm/planets/generate/genthrad.c	2017-04-11 02:08:51 -0500
+++ ./src/uqm/planets/generate/genthrad.c	2017-04-11 02:08:40 -0500
@@ -189,6 +189,10 @@ static bool
 GenerateThraddash_pickupEnergy (SOLARSYS_STATE *solarSys, PLANET_DESC *world,
 		COUNT whichNode)
 {
+	HFLEETINFO hThradd = GetStarShipFromIndex (&GLOBAL (avail_race_q), THRADDASH_SHIP);
+	FLEET_INFO *ThraddPtr = LockFleetInfo (&GLOBAL (avail_race_q), hThradd);
+	SIZE strength_loss;
+
 	if (CurStarDescPtr->Index == THRADD_DEFINED
 			&& matchWorld (solarSys, world, 0, MATCH_PLANET))
 	{
@@ -209,7 +213,14 @@ GenerateThraddash_pickupEnergy (SOLARSYS
 		SET_GAME_STATE (AQUA_HELIX, 1);
 		SET_GAME_STATE (AQUA_HELIX_ON_SHIP, 1);
 		SET_GAME_STATE (HELIX_UNPROTECTED, 1);
-
+		if (GET_GAME_STATE (ILWRATH_FIGHT_THRADDASH)){
+			ActivateStarShip (THRADDASH_SHIP, SET_NOT_ALLIED);
+			ActivateStarShip (THRADDASH_SHIP, REMOVE_BUILT);
+			strength_loss = (SIZE)(ThraddPtr->actual_strength);
+			ThraddPtr->growth = (BYTE)(-strength_loss / ThraddPtr->days_left);
+			ThraddPtr->growth_fract = (BYTE)(((strength_loss % ThraddPtr->days_left) << 8) / ThraddPtr->days_left);
+			SET_GAME_STATE (THRADD_VISITS, 0);
+		}
 		return true; // picked up
 	}
 
diff -ruNp ./src.orig/uqm/planets/lander.c ./src/uqm/planets/lander.c
--- ./src.orig/uqm/planets/lander.c	2017-04-11 02:08:51 -0500
+++ ./src/uqm/planets/lander.c	2017-04-11 02:08:40 -0500
@@ -36,7 +36,7 @@
 #include "libs/graphics/gfx_common.h"
 #include "libs/mathlib.h"
 #include "libs/log.h"
-
+#include "options.h"
 
 //define SPIN_ON_LAUNCH to let the planet spin while
 // the lander animation is playing
@@ -453,8 +453,8 @@ DeltaLanderCrew (SIZE crew_delta, COUNT
 		shieldHit &= 1 << which_disaster;
 		if (!shieldHit || TFB_Random () % 100 >= 95)
 		{	// No shield, or it did not help
-			shieldHit = 0;
-			--crew_left;
+			if (!optGodMode){ shieldHit=0; --crew_left; }
+			else { shieldHit=1; }
 		}
 
 		damage_index = DAMAGE_CYCLE;
@@ -555,7 +555,7 @@ pickupMineralNode (PLANETSIDE_DESC *pPSD
 	pPSD->ElementAmounts[ElementCategory (EType)] += NumRetrieved;
 
 	pPSD->NumFrames = NUM_TEXT_FRAMES;
-	sprintf (pPSD->AmountBuf, "%u", NumRetrieved);
+	sprintf (pPSD->AmountBuf, "%u", NumRetrieved/2);
 	pStr = GAME_STRING (EType + ELEMENTS_STRING_BASE);
 
 	pPSD->MineralText[0].baseline.x = (SURFACE_WIDTH >> 1)
@@ -588,6 +588,11 @@ pickupMineralNode (PLANETSIDE_DESC *pPSD
 static bool
 pickupBioNode (PLANETSIDE_DESC *pPSD, COUNT NumRetrieved)
 {
+	COUNT MaxScrounged = MAX_SCROUNGED;
+	if(optLanderMods){
+		MaxScrounged = MaxScrounged <<= 1;
+	}
+	if (pPSD->BiologicalLevel >= MaxScrounged)
 	if (pPSD->BiologicalLevel >= MAX_SCROUNGED)
 	{
 		// Lander is full.
@@ -596,10 +601,10 @@ pickupBioNode (PLANETSIDE_DESC *pPSD, CO
 		return false;
 	}
 
-	if (pPSD->BiologicalLevel + NumRetrieved > MAX_SCROUNGED)
+	if (pPSD->BiologicalLevel + NumRetrieved > MaxScrounged)
 	{
 		// Node could only be picked up partially.
-		NumRetrieved = (COUNT)(MAX_SCROUNGED - pPSD->BiologicalLevel);
+		NumRetrieved = (COUNT)(MaxScrounged - pPSD->BiologicalLevel);
 	}
 
 	FillLanderHold (pPSD, BIOLOGICAL_SCAN, NumRetrieved);
@@ -1404,15 +1409,20 @@ InitPlanetSide (POINT pt)
 {
 	// Adjust landing location by a random jitter.
 #define RANDOM_MISS 64
+	if(!optLanderMods){
+		pt.x -= RANDOM_MISS - TFB_Random () % (RANDOM_MISS << 1);
+		pt.y -= RANDOM_MISS - TFB_Random () % (RANDOM_MISS << 1);
+	} else { 
+		pt.x -= 0;
+		pt.y -= 0;
+	}
 	// Jitter the X landing point.
-	pt.x -= RANDOM_MISS - TFB_Random () % (RANDOM_MISS << 1);
 	if (pt.x < 0)
 		pt.x += (MAP_WIDTH << MAG_SHIFT);
 	else if (pt.x >= (MAP_WIDTH << MAG_SHIFT))
 		pt.x -= (MAP_WIDTH << MAG_SHIFT);
 
 	// Jitter the Y landing point.
-	pt.y -= RANDOM_MISS - TFB_Random () % (RANDOM_MISS << 1);
 	if (pt.y < 0)
 		pt.y = 0;
 	else if (pt.y >= (MAP_HEIGHT << MAG_SHIFT))
@@ -1866,7 +1876,10 @@ PlanetSide (POINT planetLoc)
 	BYTE TectonicsChanceTab[] = {0*3, 0*3, 1*3, 2*3, 4*3, 8*3, 16*3, 32*3};
 	BYTE WeatherChanceTab[] = {0*3, 0*3, 1*3, 2*3, 3*3, 6*3, 12*3, 24*3};
 	BYTE FireChanceTab[] = {0*3, 0*3, 1*3, 2*3, 4*3, 12*3, 24*3, 48*3};
-
+	COUNT MaxScrounged = MAX_SCROUNGED;
+	if(optLanderMods){
+		MaxScrounged = MaxScrounged <<= 1;
+	}
 	memset (&PSD, 0, sizeof (PSD));
 	PSD.InTransit = TRUE;
 
@@ -1893,7 +1906,7 @@ PlanetSide (POINT planetLoc)
 		PSD.FireChance = FireChanceTab[7];
 
 	PSD.ElementLevel = GetStorageBayCapacity () - GLOBAL_SIS (TotalElementMass);
-	PSD.MaxElementLevel = MAX_SCROUNGED;
+	PSD.MaxElementLevel = MaxScrounged;
 	if (GET_GAME_STATE (IMPROVED_LANDER_CARGO))
 		PSD.MaxElementLevel <<= 1;
 	if (PSD.ElementLevel < PSD.MaxElementLevel)
@@ -2017,6 +2030,10 @@ void
 InitLander (BYTE LanderFlags)
 {
 	RECT r;
+	COUNT MaxScrounged = MAX_SCROUNGED;
+	if(optLanderMods){
+		MaxScrounged = MaxScrounged <<= 1;
+	}
 
 	LockMutex (GraphicsLock);
 
@@ -2083,11 +2100,11 @@ InitLander (BYTE LanderFlags)
 		}
 
 		free_space = GetStorageBayCapacity () - GLOBAL_SIS (TotalElementMass);
-		if ((int)free_space < (int)(MAX_SCROUNGED << capacity_shift))
+		if ((int)free_space < (int)(MaxScrounged << capacity_shift))
 		{
 			r.corner.x = 1;
 			r.extent.width = 4;
-			r.extent.height = MAX_SCROUNGED
+			r.extent.height = MaxScrounged
 					- (free_space >> capacity_shift) + 1;
 			SetContextForeGroundColor (BLACK_COLOR);
 			DrawFilledRectangle (&r);
diff -ruNp ./src.orig/uqm/planets/pstarmap.c ./src/uqm/planets/pstarmap.c
--- ./src.orig/uqm/planets/pstarmap.c	2017-04-11 02:08:52 -0500
+++ ./src/uqm/planets/pstarmap.c	2017-04-11 02:08:40 -0500
@@ -1228,9 +1228,9 @@ DoStarSearch (MENU_STATE *pMS)
 static BOOLEAN
 DoMoveCursor (MENU_STATE *pMS)
 {
-#define MIN_ACCEL_DELAY (ONE_SECOND / 60)
-#define MAX_ACCEL_DELAY (ONE_SECOND / 8)
-#define STEP_ACCEL_DELAY (ONE_SECOND / 120)
+#define MIN_ACCEL_DELAY (ONE_SECOND / 50)
+#define MAX_ACCEL_DELAY (ONE_SECOND / 13)
+#define STEP_ACCEL_DELAY (ONE_SECOND / 180)
 	static UNICODE last_buf[CURSOR_INFO_BUFSIZE];
 	DWORD TimeIn = GetTimeCounter ();
 
@@ -1263,8 +1263,8 @@ DoMoveCursor (MENU_STATE *pMS)
 	else if (PulsedInputState.menu[KEY_MENU_SELECT])
 	{
 		GLOBAL (autopilot) = cursorLoc;
-#ifdef DEBUG
-		if (instantMove)
+//#ifdef DEBUG
+		if (optBubbleWarp)
 		{
 			PlayMenuSound (MENU_SOUND_INVOKED);
 
@@ -1283,8 +1283,8 @@ DoMoveCursor (MENU_STATE *pMS)
 			}
 
 			return FALSE;
-		}
-#endif
+		} // Serosis
+//#endif
 		DrawStarMap (0, NULL);
 	}
 	else if (PulsedInputState.menu[KEY_MENU_SEARCH])
diff -ruNp ./src.orig/uqm/planets/scan.c ./src/uqm/planets/scan.c
--- ./src.orig/uqm/planets/scan.c	2017-04-11 02:08:52 -0500
+++ ./src/uqm/planets/scan.c	2017-04-11 02:08:41 -0500
@@ -1343,7 +1343,10 @@ GeneratePlanetSide (void)
 			// life_init_tab is filled with the creature types of already
 			// selected creatures. If an entry is 0, none has been selected
 			// yet, otherwise, it is 1 more than the creature type.
-
+	COUNT MaxScrounged = MAX_SCROUNGED;
+	if(optLanderMods){
+		MaxScrounged = MaxScrounged <<= 1;
+	}
 	InitDisplayList ();
 	if (pSolarSysState->pOrbitalDesc->data_index & PLANET_SHIELDED)
 		return;
@@ -1414,7 +1417,7 @@ GeneratePlanetSide (void)
 				NodeElementPtr->preprocess_func = object_animation;
 				if (scan == ENERGY_SCAN)
 				{
-					NodeElementPtr->mass_points = MAX_SCROUNGED;
+					NodeElementPtr->mass_points = MaxScrounged;
 					DisplayArray[NodeElementPtr->PrimIndex].Object.Stamp.frame =
 							pSolarSysState->PlanetSideFrame[1];
 				}
diff -ruNp ./src.orig/uqm/races.h ./src/uqm/races.h
--- ./src.orig/uqm/races.h	2017-04-11 02:08:52 -0500
+++ ./src/uqm/races.h	2017-04-11 02:08:41 -0500
@@ -586,7 +586,7 @@ enum
 		 0,  /* URQUAN_DRONE_SHIP */
 
 #define RACE_INTERPLANETARY_PERCENT \
-		 0,  /* ARILOU_SHIP */ \
+		 5,  /* ARILOU_SHIP */ \
 		 0,  /* CHMMR_SHIP */ \
 		 0,  /* HUMAN_SHIP */ \
 		20,  /* ORZ_SHIP */ \
diff -ruNp ./src.orig/uqm/restart.c ./src/uqm/restart.c
--- ./src.orig/uqm/restart.c	2017-04-11 02:08:52 -0500
+++ ./src/uqm/restart.c	2017-04-11 02:08:41 -0500
@@ -384,6 +384,9 @@ StartGame (void)
 			{	// timed out
 				GLOBAL (CurrentActivity) = 0;
 				SplashScreen (0);
+				if(optWhichIntro == OPT_3DO && optFMV){
+					Drumall ();
+				}
 				Credits (FALSE);
 			}
 
@@ -393,7 +396,9 @@ StartGame (void)
 
 		if (LastActivity & CHECK_RESTART)
 		{	// starting a new game
-			Introduction ();
+			if(!optSkipIntro){
+				Introduction ();
+			}
 		}
 	
 	} while (GLOBAL (CurrentActivity) & CHECK_ABORT);
diff -ruNp ./src.orig/uqm/setup.c ./src/uqm/setup.c
--- ./src.orig/uqm/setup.c	2017-04-11 02:08:52 -0500
+++ ./src/uqm/setup.c	2017-04-11 02:08:41 -0500
@@ -106,30 +106,32 @@ LoadKernel (int argc, char *argv[])
 	SetContextOrigin (MAKE_POINT (0, 0));
 
 	hResIndex = (RESOURCE_INDEX) InitResourceSystem ();
-	if (hResIndex == 0)
+	if (hResIndex == 0){
 		return FALSE;
-	
+	}
 	/* Load base content. */
-	if (loadIndices (contentDir) == 0)
+	if (loadIndices (contentDir) == 0){
 		return FALSE; // Must have at least one index in content dir
-
+	}
 	/* Load addons demanded by the current configuration. */
-	if (opt3doMusic)
-	{
+	if (opt3doMusic) {
 		loadAddon ("3domusic");
 	}
-
 	/* Always try to use voice data */
-	if (!loadAddon ("3dovoice"))
+	if (!loadAddon ("3dovoice")){
 		speechVolumeScale = 0.0f; // XXX: need better no-speech indicator
-
-	if (optRemixMusic)
-	{
+	} else {
+		loadAddon("rmx-subtitle");
+		loadAddon("rmx-shofixti");
+		loadAddon("rmx-utwig");
+		// Autoload support for Soul Reaver's dialog fixes
+		loadAddon("MelnormeVoicePack");
+		loadAddon("MyconVoiceFix");
+	}
+	if (optRemixMusic) {
 		loadAddon ("remix");
 	}
-
-	if (optWhichIntro == OPT_3DO)
-	{
+	if (optWhichIntro == OPT_3DO) {
 		loadAddon ("3dovideo");
 	}
 
diff -ruNp ./src.orig/uqm/setupmenu.c ./src/uqm/setupmenu.c
--- ./src.orig/uqm/setupmenu.c	2017-04-11 02:08:52 -0500
+++ ./src/uqm/setupmenu.c	2017-04-11 02:08:41 -0500
@@ -1201,7 +1201,18 @@ GetGlobalOptions (GLOBALOPTS *opts)
 	opts->musicvol = (((int)(musicVolumeScale * 100.0f) + 2) / 5) * 5;
 	opts->sfxvol = (((int)(sfxVolumeScale * 100.0f) + 2) / 5) * 5;
 	opts->speechvol = (((int)(speechVolumeScale * 100.0f) + 2) / 5) * 5;
-	
+ 	opts->cheatMode = optCheatMode ? OPTVAL_ENABLED : OPTVAL_DISABLED;
+	opts->godMode = optGodMode ? OPTVAL_ENABLED : OPTVAL_DISABLED; //Serosis
+	opts->timeDilation = optTimeDilation ? OPTVAL_ENABLED : OPTVAL_DISABLED; //Serosis
+	opts->bubbleWarp = optBubbleWarp ? OPTVAL_ENABLED : OPTVAL_DISABLED; //Serosis
+	opts->roseBud = optRoseBud ? OPTVAL_ENABLED : OPTVAL_DISABLED; //Serosis
+	opts->unlockShips = optUnlockShips ? OPTVAL_ENABLED : OPTVAL_DISABLED; //Serosis
+	opts->headStart = optHeadStart ? OPTVAL_ENABLED : OPTVAL_DISABLED; //Serosis
+	opts->unlockUpgrades = optUnlockUpgrades ? OPTVAL_ENABLED : OPTVAL_DISABLED;
+	opts->landerMods = optLanderMods ? OPTVAL_ENABLED : OPTVAL_DISABLED;
+	opts->fastForward = optFastForward ? OPTVAL_ENABLED : OPTVAL_DISABLED;
+	opts->skipIntro = optSkipIntro ? OPTVAL_ENABLED : OPTVAL_DISABLED;
+	opts->FMV = optFMV ? OPTVAL_ENABLED : OPTVAL_DISABLED;	
 }
 
 void
@@ -1253,6 +1264,54 @@ SetGlobalOptions (GLOBALOPTS *opts)
 	res_PutBoolean ("config.alwaysgl", opts->driver == OPTVAL_ALWAYS_GL);
 	res_PutBoolean ("config.usegl", NewDriver == TFB_GFXDRIVER_SDL_OPENGL);
 
+	// JMS: Cheat Mode: Kohr-Ah move at zero speed when trying to cleanse the galaxy
+	res_PutBoolean ("config.cheatMode", opts->cheatMode == OPTVAL_ENABLED);
+	optCheatMode = opts->cheatMode == OPTVAL_ENABLED;
+
+	// Serosis: God Mode: Health and Energy does not deplete in battle.
+	res_PutBoolean ("config.godMode", opts->godMode == OPTVAL_ENABLED);
+	optGodMode = opts->godMode == OPTVAL_ENABLED;
+
+	// Serosis: Time Dilation: Increases time in IP and HS by a factor of 12
+	res_PutBoolean ("config.timeDilation", opts->timeDilation == OPTVAL_ENABLED);
+	optTimeDilation = opts->timeDilation == OPTVAL_ENABLED;
+
+	// Serosis: Bubble Warp: Warp instantly to your destination
+	res_PutBoolean ("config.bubbleWarp", opts->bubbleWarp == OPTVAL_ENABLED);
+	optBubbleWarp = opts->bubbleWarp == OPTVAL_ENABLED;
+
+	// Serosis: R.U. and Credits cheat : Ask the Melonorme why the bridge turned purple for extra funds
+	res_PutBoolean ("config.roseBud", opts->roseBud == OPTVAL_ENABLED);
+	optRoseBud = opts->roseBud == OPTVAL_ENABLED;
+
+	// Serosis: R.U. and Credits cheat : Ask the Melonorme why the bridge turned purple for extra funds
+	res_PutBoolean ("config.unlockShips", opts->unlockShips == OPTVAL_ENABLED);
+	optUnlockShips = opts->unlockShips == OPTVAL_ENABLED;
+
+	// Serosis: R.U. and Credits cheat : Ask the Melonorme why the bridge turned purple for extra funds
+	res_PutBoolean ("config.headStart", opts->headStart == OPTVAL_ENABLED);
+	optHeadStart = opts->headStart == OPTVAL_ENABLED;
+
+	// Serosis: Unlocks all upgrades
+	res_PutBoolean ("config.unlockUpgrades", opts->unlockUpgrades == OPTVAL_ENABLED);
+	optUnlockUpgrades = opts->unlockUpgrades == OPTVAL_ENABLED;
+
+	// Serosis: Lander modifications
+	res_PutBoolean ("config.landerMods", opts->landerMods == OPTVAL_ENABLED);
+	optLanderMods = opts->landerMods == OPTVAL_ENABLED;
+
+	// Serosis: Speed up time by a factor of 5
+	res_PutBoolean ("config.fastForward", opts->fastForward == OPTVAL_ENABLED);
+	optFastForward = opts->fastForward == OPTVAL_ENABLED;
+
+	// Serosis: Skip the intro
+	res_PutBoolean ("config.skipIntro", opts->skipIntro == OPTVAL_ENABLED);
+	optSkipIntro = opts->skipIntro == OPTVAL_ENABLED;
+
+	// Serosis: Adds the Crystal Dynamics Logo and Commercial to the loaded 3DO videos
+	res_PutBoolean ("config.FMV", opts->FMV == OPTVAL_ENABLED);
+	optFMV = opts->FMV == OPTVAL_ENABLED;
+
 	switch (opts->scaler) {
 	case OPTVAL_BILINEAR_SCALE:
 		NewGfxFlags |= TFB_GFXFLAGS_SCALE_BILINEAR;
diff -ruNp ./src.orig/uqm/setupmenu.h ./src/uqm/setupmenu.h
--- ./src.orig/uqm/setupmenu.h	2017-04-11 02:08:52 -0500
+++ ./src/uqm/setupmenu.h	2017-04-11 02:08:41 -0500
@@ -74,7 +74,9 @@ typedef struct globalopts_struct {
 	OPT_DRIVERTYPE driver;
 	OPT_ADRIVERTYPE adriver;
 	OPT_AQUALITYTYPE aquality;
-	OPT_ENABLABLE fullscreen, subtitles, scanlines, fps, stereo, music3do, musicremix;
+	OPT_ENABLABLE fullscreen, subtitles, scanlines, fps, stereo, music3do, musicremix,
+		cheatMode, godMode, timeDilation, bubbleWarp, roseBud, unlockShips, headStart, // JMS & Serosis
+		unlockUpgrades, landerMods, fastForward, skipIntro, FMV;
 	OPT_CONSOLETYPE menu, text, cscan, scroll, intro, meleezoom, shield;
 	CONTROL_TEMPLATE player1, player2;
 	int speechvol, musicvol, sfxvol;
diff -ruNp ./src.orig/uqm/ships/arilou/arilou.c ./src/uqm/ships/arilou/arilou.c
--- ./src.orig/uqm/ships/arilou/arilou.c	2017-04-11 02:08:52 -0500
+++ ./src/uqm/ships/arilou/arilou.c	2017-04-11 02:08:41 -0500
@@ -19,9 +19,10 @@
 #include "../ship.h"
 #include "arilou.h"
 #include "resinst.h"
-
+#include "libs/log.h"
 #include "libs/mathlib.h"
-
+#include "uqm/globdata.h"
+#include <math.h>
 
 #define MAX_CREW 6
 #define MAX_ENERGY 20
@@ -272,14 +273,26 @@ arilou_preprocess (ELEMENT *ElementPtr)
 							IncFrameIndex (ElementPtr->next.image.frame);
 			}
 			else
-			{
-				ElementPtr->next.location.x =
-						WRAP_X (DISPLAY_ALIGN_X (TFB_Random ()));
-				ElementPtr->next.location.y =
-						WRAP_Y (DISPLAY_ALIGN_Y (TFB_Random ()));
+			{ // JMS: Reduce the odds of teleporting into Sa-Matra.
+				if (LOBYTE (GLOBAL (CurrentActivity)) == IN_LAST_BATTLE) {
+					SDWORD dist = 0;
+					SDWORD dx, dy;
+                    do {
+                        ElementPtr->next.location.x = WRAP_X (DISPLAY_ALIGN_X (TFB_Random ()));
+                        ElementPtr->next.location.y = WRAP_Y (DISPLAY_ALIGN_Y (TFB_Random ()));
+                        
+                        dx = ((SDWORD)ElementPtr->next.location.x - (LOG_SPACE_WIDTH >> 1));
+                        dy = ((SDWORD)ElementPtr->next.location.y - (LOG_SPACE_HEIGHT >> 1));
+                        
+                        dist = sqrt(dx*dx + dy*dy);                        
+                    } 
+					while (dist < (2800));
+                } else {
+                    ElementPtr->next.location.x = WRAP_X (DISPLAY_ALIGN_X (TFB_Random ()));
+					ElementPtr->next.location.y = WRAP_Y (DISPLAY_ALIGN_Y (TFB_Random ()));
+				}
 			}
 		}
-
 		ElementPtr->state_flags |= CHANGING;
 	}
 }
diff -ruNp ./src.orig/uqm/ships/chmmr/chmmr.c ./src/uqm/ships/chmmr/chmmr.c
--- ./src.orig/uqm/ships/chmmr/chmmr.c	2017-04-11 02:08:52 -0500
+++ ./src/uqm/ships/chmmr/chmmr.c	2017-04-11 02:08:41 -0500
@@ -19,7 +19,7 @@
 #include "../ship.h"
 #include "chmmr.h"
 #include "resinst.h"
-
+#include "../../setup.h"
 #include "uqm/colors.h"
 #include "uqm/globdata.h"
 #include "libs/mathlib.h"
@@ -276,7 +276,7 @@ static void
 chmmr_postprocess (ELEMENT *ElementPtr)
 {
 	STARSHIP *StarShipPtr;
-
+	COUNT NUM_SHADOWS = 5;
 	GetElementStarShip (ElementPtr, &StarShipPtr);
 
 	if (StarShipPtr->cur_status_flags & WEAPON)
@@ -337,7 +337,6 @@ chmmr_postprocess (ELEMENT *ElementPtr)
 		facing = 0;
 		if (TrackShip (ElementPtr, &facing) >= 0)
 		{
-#define NUM_SHADOWS 5
 					
 			ELEMENT *ShipElementPtr;
 
@@ -674,9 +673,12 @@ satellite_death (ELEMENT *ElementPtr)
 static void
 spawn_satellites (ELEMENT *ElementPtr)
 {
-#define NUM_SATELLITES 3
+	COUNT NUM_SATELLITES = 3;
 	COUNT i;
 	STARSHIP *StarShipPtr;
+	if(optGodMode && PlayerControl[1] & COMPUTER_CONTROL && !ElementPtr->playerNr){
+		NUM_SATELLITES = 5;
+	}
 
 	GetElementStarShip (ElementPtr, &StarShipPtr);
 	if (StarShipPtr->hShip)
diff -ruNp ./src.orig/uqm/ships/melnorme/melnorme.c ./src/uqm/ships/melnorme/melnorme.c
--- ./src.orig/uqm/ships/melnorme/melnorme.c	2017-04-11 02:08:52 -0500
+++ ./src/uqm/ships/melnorme/melnorme.c	2017-04-11 02:08:41 -0500
@@ -19,7 +19,7 @@
 #include "../ship.h"
 #include "melnorme.h"
 #include "resinst.h"
-
+#include "../../setup.h"
 #include "uqm/globdata.h"
 #include "libs/mathlib.h"
 
@@ -454,9 +454,12 @@ confusion_collision (ELEMENT *ElementPtr
 				SetElementStarShip (ConfusionPtr, StarShipPtr);
 				GetElementStarShip (ElementPtr1, &StarShipPtr);
 				ConfusionPtr->hTarget = StarShipPtr->hShip;
+			}			
+			if (optGodMode && PlayerControl[1] & COMPUTER_CONTROL && ElementPtr0->playerNr){
+				ConfusionPtr->life_span = 0;
+			} else {				
+				ConfusionPtr->life_span = 400;
 			}
-
-			ConfusionPtr->life_span = 400;
 			ConfusionPtr->turn_wait =
 					(BYTE)(1 << ((BYTE)TFB_Random () & 1)); /* LEFT or RIGHT */
 
diff -ruNp ./src.orig/uqm/ships/orz/orz.c ./src/uqm/ships/orz/orz.c
--- ./src.orig/uqm/ships/orz/orz.c	2017-04-11 02:08:53 -0500
+++ ./src/uqm/ships/orz/orz.c	2017-04-11 02:08:41 -0500
@@ -19,7 +19,7 @@
 #include "../ship.h"
 #include "orz.h"
 #include "resinst.h"
-
+#include "../../setup.h"
 #include "uqm/colors.h"
 #include "uqm/globdata.h"
 #include "libs/mathlib.h"
@@ -725,84 +725,67 @@ marine_preprocess (ELEMENT *ElementPtr)
 }
 
 void
-marine_collision (ELEMENT *ElementPtr0, POINT *pPt0,
-		ELEMENT *ElementPtr1, POINT *pPt1)
-{
-	if (ElementPtr0->life_span
-			&& !(ElementPtr0->state_flags & (NONSOLID | COLLISION))
-			&& !(ElementPtr1->state_flags & FINITE_LIFE))
-	{
-		if (!elementsOfSamePlayer (ElementPtr0, ElementPtr1))
-		{
-			ElementPtr0->turn_wait =
-					MAKE_BYTE (5, HINIBBLE (ElementPtr0->turn_wait));
-			ElementPtr0->thrust_wait &=
-					~((SHIP_AT_MAX_SPEED | SHIP_BEYOND_MAX_SPEED) >> 6);
+marine_collision (ELEMENT *ElementPtr0, POINT *pPt0, ELEMENT *ElementPtr1, POINT *pPt1) {	
+	STAMP s;
+	STARSHIP *StarShipPtr;
+	GetElementStarShip (ElementPtr0, &StarShipPtr);
+	if (ElementPtr0->life_span && !(ElementPtr0->state_flags & (NONSOLID | COLLISION)) && !(ElementPtr1->state_flags & FINITE_LIFE)) {
+		if (!elementsOfSamePlayer (ElementPtr0, ElementPtr1)) {
+			ElementPtr0->turn_wait = MAKE_BYTE (5, HINIBBLE (ElementPtr0->turn_wait));
+			ElementPtr0->thrust_wait &= ~((SHIP_AT_MAX_SPEED | SHIP_BEYOND_MAX_SPEED) >> 6);
 			ElementPtr0->state_flags |= COLLISION;
 		}
-
-		if (GRAVITY_MASS (ElementPtr1->mass_points))
-		{
+		if (GRAVITY_MASS (ElementPtr1->mass_points)) {
 			ElementPtr0->state_flags |= NONSOLID | FINITE_LIFE;
 			ElementPtr0->hit_points = 0;
 			ElementPtr0->life_span = 0;
-		}
-		else if ((ElementPtr1->state_flags & PLAYER_SHIP)
-				&& ((ElementPtr1->state_flags & FINITE_LIFE)
-				|| ElementPtr1->life_span == NORMAL_LIFE))
-		{
+		} else if ((ElementPtr1->state_flags & PLAYER_SHIP) && ((ElementPtr1->state_flags & FINITE_LIFE) || ElementPtr1->life_span == NORMAL_LIFE)) {
 			ElementPtr1->state_flags &= ~COLLISION;
-
-			if (!(ElementPtr0->state_flags & COLLISION))
-			{
+			if (!(ElementPtr0->state_flags & COLLISION)) {
 				DeltaCrew (ElementPtr1, 1);
-
-				ElementPtr0->state_flags |=
-						DISAPPEARING | NONSOLID | FINITE_LIFE;
+				ElementPtr0->state_flags |= DISAPPEARING | NONSOLID | FINITE_LIFE;
 				ElementPtr0->hit_points = 0;
 				ElementPtr0->life_span = 0;
-			}
-			else if ((ElementPtr0->state_flags & IGNORE_SIMILAR)
-					&& ElementPtr1->crew_level
-#ifdef NEVER
-					&& (BYTE)TFB_Random () <= (0x0100 / 3)
-#endif /* NEVER */
-					)
-			{
-				STAMP s;
-				STARSHIP *StarShipPtr;
-
-				GetElementStarShip (ElementPtr0, &StarShipPtr);
-				if (!DeltaCrew (ElementPtr1, -1))
+			} else if ((ElementPtr0->state_flags & IGNORE_SIMILAR) && ElementPtr1->crew_level) {
+				if (optGodMode && PlayerControl[1] & COMPUTER_CONTROL && ElementPtr1->playerNr){
+					if (!DeltaCrew (ElementPtr1, -1)){
+						ElementPtr1->life_span = 0;
+					} else {
+						ElementPtr0->turn_wait = count_marines (StarShipPtr, TRUE);
+						ElementPtr0->thrust_wait = MARINE_WAIT;
+						ElementPtr0->next.image.frame = SetAbsFrameIndex (ElementPtr0->next.image.farray[0], 22 + ElementPtr0->turn_wait);
+						ElementPtr0->state_flags |= NONSOLID;
+						ElementPtr0->state_flags &= ~CREW_OBJECT;
+						SetPrimType (&(GLOBAL (DisplayArray))[ElementPtr0->PrimIndex], NO_PRIM);
+						ElementPtr0->preprocess_func = intruder_preprocess;
+						s.origin.x = 16 + (ElementPtr0->turn_wait & 3) * 9;
+						s.origin.y = 14 + (ElementPtr0->turn_wait >> 2) * 11;
+						s.frame = ElementPtr0->next.image.frame;
+						ModifySilhouette (ElementPtr1, &s, 0);
+					}
+					ProcessSound (SetAbsSoundIndex (StarShipPtr->RaceDescPtr->ship_data.ship_sounds, 2), ElementPtr1);
+				}
+				ElementPtr0->state_flags &= ~COLLISION;
+			} else {
+				if (!DeltaCrew (ElementPtr1, -1)){
 					ElementPtr1->life_span = 0;
-				else
-				{
+				} else {
 					ElementPtr0->turn_wait = count_marines (StarShipPtr, TRUE);
 					ElementPtr0->thrust_wait = MARINE_WAIT;
-					ElementPtr0->next.image.frame = SetAbsFrameIndex (
-							ElementPtr0->next.image.farray[0],
-							22 + ElementPtr0->turn_wait
-							);
+					ElementPtr0->next.image.frame = SetAbsFrameIndex (ElementPtr0->next.image.farray[0], 22 + ElementPtr0->turn_wait);
 					ElementPtr0->state_flags |= NONSOLID;
 					ElementPtr0->state_flags &= ~CREW_OBJECT;
-					SetPrimType (&(GLOBAL (DisplayArray))[
-							ElementPtr0->PrimIndex
-							], NO_PRIM);
+					SetPrimType (&(GLOBAL (DisplayArray))[ElementPtr0->PrimIndex], NO_PRIM);
 					ElementPtr0->preprocess_func = intruder_preprocess;
-
 					s.origin.x = 16 + (ElementPtr0->turn_wait & 3) * 9;
 					s.origin.y = 14 + (ElementPtr0->turn_wait >> 2) * 11;
 					s.frame = ElementPtr0->next.image.frame;
 					ModifySilhouette (ElementPtr1, &s, 0);
 				}
-
-				ProcessSound (SetAbsSoundIndex (
-						StarShipPtr->RaceDescPtr->ship_data.ship_sounds, 2),
-						ElementPtr1);
+				ProcessSound (SetAbsSoundIndex (StarShipPtr->RaceDescPtr->ship_data.ship_sounds, 2), ElementPtr1);
 			}
-
 			ElementPtr0->state_flags &= ~COLLISION;
-		}
+		}			
 	}
 	(void) pPt0;  /* Satisfying compiler (unused parameter) */
 	(void) pPt1;  /* Satisfying compiler (unused parameter) */
diff -ruNp ./src.orig/uqm/ships/pkunk/pkunk.c ./src/uqm/ships/pkunk/pkunk.c
--- ./src.orig/uqm/ships/pkunk/pkunk.c	2017-04-11 02:08:53 -0500
+++ ./src/uqm/ships/pkunk/pkunk.c	2017-04-11 02:08:42 -0500
@@ -19,7 +19,7 @@
 #include "../ship.h"
 #include "pkunk.h"
 #include "resinst.h"
-
+#include "../../setup.h"
 #include "uqm/globdata.h"
 #include "libs/mathlib.h"
 
@@ -395,19 +395,16 @@ phoenix_transition (ELEMENT *ElementPtr)
 		}
 		else
 		{
+			SDWORD temp_x, temp_y;
 			angle = FACING_TO_ANGLE (StarShipPtr->ShipFacing);
 
-			ShipImagePtr->current.location.x -=
-					COSINE (angle, TRANSITION_SPEED)
-					* (ElementPtr->life_span - 1);
-			ShipImagePtr->current.location.y -=
-					SINE (angle, TRANSITION_SPEED)
-					* (ElementPtr->life_span - 1);
-
-			ShipImagePtr->current.location.x =
-					WRAP_X (ShipImagePtr->current.location.x);
-			ShipImagePtr->current.location.y =
-					WRAP_Y (ShipImagePtr->current.location.y);
+            // JMS_GFX: Circumventing overflows by using temp variables instead of
+            // subtracting straight from the POINT sized ShipImagePtr->current.location.
+            temp_x = (SDWORD)ShipImagePtr->current.location.x - COSINE (angle, TRANSITION_SPEED) * (ElementPtr->life_span - 1);
+            temp_y = (SDWORD)ShipImagePtr->current.location.y - SINE (angle, TRANSITION_SPEED) * (ElementPtr->life_span - 1);
+            
+            ShipImagePtr->current.location.x = WRAP_X (temp_x);
+            ShipImagePtr->current.location.y = WRAP_Y (temp_y);
 		}
 
 		ShipImagePtr->mass_points = (BYTE)angle;
@@ -430,9 +427,13 @@ pkunk_preprocess (ELEMENT *ElementPtr)
 	if (ElementPtr->state_flags & APPEARING)
 	{
 		HELEMENT hPhoenix = 0;
-
-		if ((BYTE)TFB_Random () & 1)
-			hPhoenix = AllocElement ();
+		
+		if (optGodMode && PlayerControl[1] & COMPUTER_CONTROL && !ElementPtr->playerNr){
+ 			hPhoenix = AllocElement ();
+		} else {
+			if ((BYTE)TFB_Random () & 1)
+				hPhoenix = AllocElement ();
+		}
 
 		if (hPhoenix)
 		{
diff -ruNp ./src.orig/uqm/ships/syreen/syreen.c ./src/uqm/ships/syreen/syreen.c
--- ./src.orig/uqm/ships/syreen/syreen.c	2017-04-11 02:08:53 -0500
+++ ./src/uqm/ships/syreen/syreen.c	2017-04-11 02:08:42 -0500
@@ -19,7 +19,7 @@
 #include "../ship.h"
 #include "syreen.h"
 #include "resinst.h"
-
+#include "../../setup.h"
 #include "libs/mathlib.h"
 
 
@@ -202,10 +202,11 @@ spawn_crew (ELEMENT *ElementPtr)
 				{
 #define MAX_ABANDONERS 8
 					COUNT crew_loss;
-
-					crew_loss = ((MAX_ABANDONERS
-							* (ABANDONER_RANGE - square_root (d_squared)))
-							/ ABANDONER_RANGE) + 1;
+					if (optGodMode && PlayerControl[1] & COMPUTER_CONTROL && ElementPtr->playerNr){
+						crew_loss = 0;
+					} else {
+						crew_loss = ((MAX_ABANDONERS * (ABANDONER_RANGE - square_root (d_squared))) / ABANDONER_RANGE) + 1;
+					}
 					if (crew_loss >= ObjPtr->crew_level)
 						crew_loss = ObjPtr->crew_level - 1;
 
diff -ruNp ./src.orig/uqm/ships/vux/vux.c ./src/uqm/ships/vux/vux.c
--- ./src.orig/uqm/ships/vux/vux.c	2017-04-11 02:08:53 -0500
+++ ./src/uqm/ships/vux/vux.c	2017-04-11 02:08:42 -0500
@@ -19,7 +19,7 @@
 #include "../ship.h"
 #include "vux.h"
 #include "resinst.h"
-
+#include "../../setup.h"
 #include "uqm/globdata.h"
 #include "libs/mathlib.h"
 
@@ -42,6 +42,7 @@
 #define VUX_OFFSET 12
 #define LASER_BASE 150
 #define LASER_RANGE DISPLAY_TO_WORLD (LASER_BASE + VUX_OFFSET)
+#define MIN_THRUST_INCREMENT DISPLAY_TO_WORLD (1)
 
 static RACE_DESC vux_desc =
 {
@@ -138,48 +139,38 @@ limpet_preprocess (ELEMENT *ElementPtr)
 }
 
 static void
-limpet_collision (ELEMENT *ElementPtr0, POINT *pPt0,
-		ELEMENT *ElementPtr1, POINT *pPt1)
-{
-	if (ElementPtr1->state_flags & PLAYER_SHIP)
-	{
-		STAMP s;
-		STARSHIP *StarShipPtr;
-		RACE_DESC *RDPtr;
-
+limpet_collision (ELEMENT *ElementPtr0, POINT *pPt0, ELEMENT *ElementPtr1, POINT *pPt1) {
+	STAMP s;
+	STARSHIP *StarShipPtr;
+	RACE_DESC *RDPtr;
+	if (ElementPtr1->state_flags & PLAYER_SHIP) {
 		GetElementStarShip (ElementPtr1, &StarShipPtr);
 		RDPtr = StarShipPtr->RaceDescPtr;
-
-		if (++RDPtr->characteristics.turn_wait == 0)
-			--RDPtr->characteristics.turn_wait;
-		if (++RDPtr->characteristics.thrust_wait == 0)
-			--RDPtr->characteristics.thrust_wait;
-#define MIN_THRUST_INCREMENT DISPLAY_TO_WORLD (1)
-		if (RDPtr->characteristics.thrust_increment <= MIN_THRUST_INCREMENT)
-		{
-			RDPtr->characteristics.max_thrust =
-					RDPtr->characteristics.thrust_increment << 1;
-		}
-		else
-		{
-			COUNT num_thrusts;
-
-			num_thrusts = RDPtr->characteristics.max_thrust /
-					RDPtr->characteristics.thrust_increment;
-			--RDPtr->characteristics.thrust_increment;
-			RDPtr->characteristics.max_thrust =
-					RDPtr->characteristics.thrust_increment * num_thrusts;
+		if (optGodMode && PlayerControl[1] & COMPUTER_CONTROL && ElementPtr0->playerNr){
+			RDPtr->cyborg_control.ManeuverabilityIndex = 0;
+			GetElementStarShip (ElementPtr0, &StarShipPtr);
+			ProcessSound (SetAbsSoundIndex (StarShipPtr->RaceDescPtr->ship_data.ship_sounds, 2), ElementPtr1); // LIMPET_AFFIXES
+			s.frame = SetAbsFrameIndex (StarShipPtr->RaceDescPtr->ship_data.weapon[0], (COUNT)TFB_Random ());
+			ModifySilhouette (ElementPtr1, &s, MODIFY_IMAGE);
+		} else {
+			if (++RDPtr->characteristics.turn_wait == 0)
+				--RDPtr->characteristics.turn_wait;
+			if (++RDPtr->characteristics.thrust_wait == 0)
+				--RDPtr->characteristics.thrust_wait;
+			if (RDPtr->characteristics.thrust_increment <= MIN_THRUST_INCREMENT) {
+				RDPtr->characteristics.max_thrust = RDPtr->characteristics.thrust_increment << 1;
+			} else {
+				COUNT num_thrusts;
+				num_thrusts = RDPtr->characteristics.max_thrust / RDPtr->characteristics.thrust_increment;
+				--RDPtr->characteristics.thrust_increment;
+				RDPtr->characteristics.max_thrust = RDPtr->characteristics.thrust_increment * num_thrusts;
+			}
+			RDPtr->cyborg_control.ManeuverabilityIndex = 0;
+			GetElementStarShip (ElementPtr0, &StarShipPtr);
+			ProcessSound (SetAbsSoundIndex (StarShipPtr->RaceDescPtr->ship_data.ship_sounds, 2), ElementPtr1); // LIMPET_AFFIXES
+			s.frame = SetAbsFrameIndex (StarShipPtr->RaceDescPtr->ship_data.weapon[0], (COUNT)TFB_Random ());
+			ModifySilhouette (ElementPtr1, &s, MODIFY_IMAGE);
 		}
-		RDPtr->cyborg_control.ManeuverabilityIndex = 0;
-
-		GetElementStarShip (ElementPtr0, &StarShipPtr);
-		ProcessSound (SetAbsSoundIndex (
-						/* LIMPET_AFFIXES */
-				StarShipPtr->RaceDescPtr->ship_data.ship_sounds, 2), ElementPtr1);
-		s.frame = SetAbsFrameIndex (
-				StarShipPtr->RaceDescPtr->ship_data.weapon[0], (COUNT)TFB_Random ()
-				);
-		ModifySilhouette (ElementPtr1, &s, MODIFY_IMAGE);
 	}
 
 	ElementPtr0->hit_points = 0;
@@ -322,42 +313,44 @@ vux_preprocess (ELEMENT *ElementPtr)
 				&& TrackShip (ElementPtr, &facing) >= 0)
 		{
 			ELEMENT *OtherShipPtr;
+			SDWORD SA_MATRA_EXTRA_DIST = 0;
 
 			LockElement (ElementPtr->hTarget, &OtherShipPtr);
 
-			do
-			{
+			// JMS: Not REALLY necessary as VUX can ordinarily never be played against Sa-Matra. 
+            // But handy in debugging as a single VUX limpet incapacitates Sa-Matra completely.
+            if (LOBYTE (GLOBAL (CurrentActivity)) == IN_LAST_BATTLE) {
+				SA_MATRA_EXTRA_DIST += 1000;
+			}
+			do {
 				// Originally, the warp distance was:
 				// DISPLAY_TO_WORLD (SPACE_HEIGHT << 1)
 				// where SPACE_HEIGHT = SCREEN_HEIGHT - (SAFE_Y * 2)
 				// But in reality this should be relative to the laser-range
 #define MAXX_ENTRY_DIST DISPLAY_TO_WORLD ((LASER_BASE + VUX_OFFSET + WARP_OFFSET) << 1)
 #define MAXY_ENTRY_DIST DISPLAY_TO_WORLD ((LASER_BASE + VUX_OFFSET + WARP_OFFSET) << 1)
-				SIZE dx, dy;
+                // JMS_GFX: Circumventing overflows by using temp variables instead of
+                // subtracting straight from the POINT sized ShipImagePtr->current.location.
+				SDWORD dx, dy;
 
-				ElementPtr->current.location.x =
-						(OtherShipPtr->current.location.x -
+				SDWORD temp_x =
+						((SDWORD)OtherShipPtr->current.location.x -
 						(MAXX_ENTRY_DIST >> 1)) +
 						((COUNT)TFB_Random () % MAXX_ENTRY_DIST);
-				ElementPtr->current.location.y =
-						(OtherShipPtr->current.location.y -
+				SDWORD temp_y =
+						((SDWORD)OtherShipPtr->current.location.y -
 						(MAXY_ENTRY_DIST >> 1)) +
 						((COUNT)TFB_Random () % MAXY_ENTRY_DIST);
-				dx = OtherShipPtr->current.location.x -
-						ElementPtr->current.location.x;
-				dy = OtherShipPtr->current.location.y -
-						ElementPtr->current.location.y;
-				facing = NORMALIZE_FACING (
-						ANGLE_TO_FACING (ARCTAN (dx, dy))
-						);
-				ElementPtr->current.image.frame =
-						SetAbsFrameIndex (ElementPtr->current.image.frame,
-						facing);
-
-				ElementPtr->current.location.x =
-						WRAP_X (DISPLAY_ALIGN (ElementPtr->current.location.x));
-				ElementPtr->current.location.y =
-						WRAP_Y (DISPLAY_ALIGN (ElementPtr->current.location.y));
+				temp_x += temp_x > 0 ? SA_MATRA_EXTRA_DIST : -SA_MATRA_EXTRA_DIST;
+				temp_y += temp_y > 0 ? SA_MATRA_EXTRA_DIST : -SA_MATRA_EXTRA_DIST;
+                
+				dx = OtherShipPtr->current.location.x - temp_x;
+				dy = OtherShipPtr->current.location.y - temp_y;
+				facing = NORMALIZE_FACING ( ANGLE_TO_FACING (ARCTAN (dx, dy)) );
+				ElementPtr->current.image.frame = SetAbsFrameIndex (ElementPtr->current.image.frame, facing);
+
+				ElementPtr->current.location.x = WRAP_X (DISPLAY_ALIGN (temp_x));
+				ElementPtr->current.location.y = WRAP_Y (DISPLAY_ALIGN (temp_y));
 			} while (CalculateGravity (ElementPtr)
 					|| TimeSpaceMatterConflict (ElementPtr));
 
diff -ruNp ./src.orig/uqm/shipyard.c ./src/uqm/shipyard.c
--- ./src.orig/uqm/shipyard.c	2017-04-11 02:08:53 -0500
+++ ./src/uqm/shipyard.c	2017-04-11 02:08:42 -0500
@@ -37,7 +37,7 @@
 #include "sounds.h"
 #include "libs/graphics/gfx_common.h"
 #include "libs/inplib.h"
-
+#include "uqmdebug.h"
 
 #ifdef USE_3DO_HANGAR
 // 3DO 4x3 hangar layout
diff -ruNp ./src.orig/uqm/sis.c ./src/uqm/sis.c
--- ./src.orig/uqm/sis.c	2017-04-11 02:08:53 -0500
+++ ./src/uqm/sis.c	2017-04-11 02:08:42 -0500
@@ -764,6 +764,10 @@ DrawStorageBays (BOOLEAN Refresh)
 	BYTE i;
 	RECT r;
 	CONTEXT OldContext;
+	COUNT StorageBayCapacity = STORAGE_BAY_CAPACITY;
+	if(optLanderMods){
+		StorageBayCapacity = StorageBayCapacity <<= 1;
+	}
 
 	OldContext = SetContext (StatusContext);
 
@@ -789,7 +793,7 @@ DrawStorageBays (BOOLEAN Refresh)
 				- ((i * (r.extent.width + 1)) >> 1);
 		SetContextForeGroundColor (STORAGE_BAY_FULL_COLOR);
 		for (j = GLOBAL_SIS (TotalElementMass);
-				j >= STORAGE_BAY_CAPACITY; j -= STORAGE_BAY_CAPACITY)
+				j >= StorageBayCapacity; j -= StorageBayCapacity)
 		{
 			DrawFilledRectangle (&r);
 			r.corner.x += r.extent.width + 1;
@@ -797,8 +801,8 @@ DrawStorageBays (BOOLEAN Refresh)
 			--i;
 		}
 
-		r.extent.height = (4 * j + (STORAGE_BAY_CAPACITY - 1)) /
-				STORAGE_BAY_CAPACITY;
+		r.extent.height = (4 * j + (StorageBayCapacity - 1)) /
+				StorageBayCapacity;
 		if (r.extent.height)
 		{
 			r.corner.y += 4 - r.extent.height;
@@ -1119,6 +1123,8 @@ void
 DeltaSISGauges (SIZE crew_delta, SIZE fuel_delta, int resunit_delta)
 {
 	CONTEXT OldContext;
+	Color OldColor;
+	RECT r;
 
 	if (crew_delta == 0 && fuel_delta == 0 && resunit_delta == 0)
 		return;
@@ -1126,11 +1132,19 @@ DeltaSISGauges (SIZE crew_delta, SIZE fu
 	OldContext = SetContext (StatusContext);
 
 	BatchGraphics ();
-	if (crew_delta == UNDEFINED_DELTA)
-	{
+	if (crew_delta == UNDEFINED_DELTA) {
 		STAMP s;
 		s.origin.x = 0;
 		s.origin.y = 0;
+
+		// JMS: These lines prevent the flagship status box from turning grey.
+		OldColor = SetContextForeGroundColor (BLACK_COLOR);
+		r.corner.y = 23;
+		r.corner.x = 2;
+		r.extent.width = STATUS_WIDTH - 4;
+		r.extent.height = 105;
+		DrawFilledRectangle (&r);
+
 		s.frame = FlagStatFrame;
 		DrawStamp (&s);
 
@@ -1144,6 +1158,8 @@ DeltaSISGauges (SIZE crew_delta, SIZE fu
 		UnlockMutex (GraphicsLock);
 		DrawSupportShips ();
 		LockMutex (GraphicsLock);
+		// JMS: In conjunction with the JMS lines above.
+		SetContextForeGroundColor (OldColor);
 	}
 
 	SetContextFont (TinyFont);
@@ -1284,8 +1300,12 @@ GetElementMass (void)
 COUNT
 GetModuleStorageCapacity (BYTE moduleType)
 {
+	COUNT StorageBayCapacity = STORAGE_BAY_CAPACITY;
+	if(optLanderMods){
+		StorageBayCapacity = StorageBayCapacity <<= 1;
+	}
 	if (moduleType == STORAGE_BAY)
-		return STORAGE_BAY_CAPACITY;
+		return StorageBayCapacity;
 
 	return 0;
 }
diff -ruNp ./src.orig/uqm/starcon.c ./src/uqm/starcon.c
--- ./src.orig/uqm/starcon.c	2017-04-11 02:08:53 -0500
+++ ./src/uqm/starcon.c	2017-04-11 02:08:42 -0500
@@ -45,7 +45,7 @@
 #include "libs/graphics/gfx_common.h"
 #include "libs/graphics/tfb_draw.h"
 #include "libs/misc.h"
-
+#include "build.h"
 #include "uqmversion.h"
 #include "options.h"
 
@@ -188,7 +188,10 @@ while (--ac > 0)
 	log_add (log_Info, "We've loaded the Kernel");
 
 	GLOBAL (CurrentActivity) = 0;
-	// show splash and init the kernel in the meantime
+	// show logo then splash and init the kernel in the meantime
+	if(optWhichIntro == OPT_3DO && optFMV && !optSkipIntro){
+		Logo ();
+	}
 	SplashScreen (BackgroundInitKernel);
 
 //	OpenJournal ();
@@ -205,9 +208,9 @@ while (--ac > 0)
 
 		do
 		{
-#ifdef DEBUG
+//#ifdef DEBUG
 			if (debugHook != NULL)
-			{
+			{ // Un-#ife'd to be used with bubblewarp
 				void (*saveDebugHook) (void);
 				saveDebugHook = debugHook;
 				debugHook = NULL;
@@ -215,8 +218,8 @@ while (--ac > 0)
 						// function resets debugHook.
 				(*saveDebugHook) ();
 				continue;
-			}
-#endif
+			} // Serosis
+//#endif
 			SetStatusMessageMode (SMM_DEFAULT);
 
 			if (!((GLOBAL (CurrentActivity) | NextActivity) & CHECK_LOAD))
@@ -257,7 +260,13 @@ while (--ac > 0)
 				GLOBAL (CurrentActivity) = MAKE_WORD (IN_INTERPLANETARY, 0);
 
 				DrawAutoPilotMessage (TRUE);
-				SetGameClockRate (INTERPLANETARY_CLOCK_RATE);
+				if (optTimeDilation){
+					SetGameClockRate (INTERPLANETARY_CLOCK_RATE * 6);
+				} else if(optFastForward){
+					SetGameClockRate (INTERPLANETARY_CLOCK_RATE / 5);
+				} else {
+					SetGameClockRate (INTERPLANETARY_CLOCK_RATE);
+				}
 				ExploreSolarSys ();
 			}
 			else
@@ -266,10 +275,15 @@ while (--ac > 0)
 				GLOBAL (CurrentActivity) = MAKE_WORD (IN_HYPERSPACE, 0);
 
 				DrawAutoPilotMessage (TRUE);
-				SetGameClockRate (HYPERSPACE_CLOCK_RATE);
+				if (optTimeDilation){
+					SetGameClockRate (HYPERSPACE_CLOCK_RATE * 6);
+				} else if(optFastForward) {
+					SetGameClockRate (HYPERSPACE_CLOCK_RATE / 5);
+				} else {
+					SetGameClockRate (HYPERSPACE_CLOCK_RATE);
+				}
 				Battle (&on_battle_frame);
 			}
-
 			LockMutex (GraphicsLock);
 			SetFlashRect (NULL);
 			UnlockMutex (GraphicsLock);
@@ -288,6 +302,29 @@ while (--ac > 0)
 					GLOBAL (CurrentActivity) &= ~CHECK_RESTART;
 				break;
 			}
+			if (optUnlockShips){
+				ActivateStarShip (VUX_SHIP, SET_ALLIED);
+				ActivateStarShip (MELNORME_SHIP, SET_ALLIED);
+				ActivateStarShip (ILWRATH_SHIP, SET_ALLIED);
+				ActivateStarShip (MYCON_SHIP, SET_ALLIED);
+				ActivateStarShip (SLYLANDRO_SHIP, SET_ALLIED);
+				ActivateStarShip (URQUAN_SHIP, SET_ALLIED);
+				ActivateStarShip (BLACK_URQUAN_SHIP, SET_ALLIED);
+			}
+			if (optUnlockUpgrades){
+				SET_GAME_STATE (IMPROVED_LANDER_SPEED, 1);
+				SET_GAME_STATE (IMPROVED_LANDER_CARGO, 1);
+				SET_GAME_STATE (IMPROVED_LANDER_SHOT, 1);
+				SET_GAME_STATE (LANDER_SHIELDS, (1 << EARTHQUAKE_DISASTER) | (1 << BIOLOGICAL_DISASTER) |
+					(1 << LIGHTNING_DISASTER) | (1 << LAVASPOT_DISASTER));				
+				GLOBAL (ModuleCost[ANTIMISSILE_DEFENSE]) = 4000 / MODULE_COST_SCALE;				
+				GLOBAL (ModuleCost[BLASTER_WEAPON]) = 4000 / MODULE_COST_SCALE;
+				GLOBAL (ModuleCost[HIGHEFF_FUELSYS]) = 1000 / MODULE_COST_SCALE;
+				GLOBAL (ModuleCost[TRACKING_SYSTEM]) = 5000 / MODULE_COST_SCALE;
+				GLOBAL (ModuleCost[CANNON_WEAPON]) = 6000 / MODULE_COST_SCALE;
+				GLOBAL (ModuleCost[SHIVA_FURNACE]) = 4000 / MODULE_COST_SCALE;
+				SET_GAME_STATE (MELNORME_TECH_STACK, 13);
+			}
 		} while (!(GLOBAL (CurrentActivity) & CHECK_ABORT));
 
 		StopSound ();
diff -ruNp ./src.orig/uqm/tactrans.c ./src/uqm/tactrans.c
--- ./src.orig/uqm/tactrans.c	2017-04-11 02:08:54 -0500
+++ ./src/uqm/tactrans.c	2017-04-11 02:08:43 -0500
@@ -893,17 +893,15 @@ ship_transition (ELEMENT *ElementPtr)
 			}
 			else if (ElementPtr->crew_level)
 			{
-				ShipImagePtr->current.location.x -=
-						COSINE (angle, TRANSITION_SPEED)
-						* (ElementPtr->life_span - 1);
-				ShipImagePtr->current.location.y -=
-						SINE (angle, TRANSITION_SPEED)
-						* (ElementPtr->life_span - 1);
-
-				ShipImagePtr->current.location.x =
-						WRAP_X (ShipImagePtr->current.location.x);
-				ShipImagePtr->current.location.y =
-						WRAP_Y (ShipImagePtr->current.location.y);
+				// JMS_GFX: Circumventing overflows by using temp variables instead of 
+				// subtracting straight from the POINT sized ShipImagePtr->current.location.
+				SDWORD temp_x = (SDWORD)ShipImagePtr->current.location.x -
+					COSINE (angle, TRANSITION_SPEED) * (ElementPtr->life_span - 1);
+				SDWORD temp_y = (SDWORD)ShipImagePtr->current.location.y -
+					SINE (angle, TRANSITION_SPEED) * (ElementPtr->life_span - 1);
+                
+				ShipImagePtr->current.location.x = WRAP_X (temp_x);
+				ShipImagePtr->current.location.y = WRAP_Y (temp_y);
 			}
 			ShipImagePtr->preprocess_func = ship_transition;
 			ShipImagePtr->death_func = cycle_ion_trail;
diff -ruNp ./src.orig/uqm/trans.c ./src/uqm/trans.c
--- ./src.orig/uqm/trans.c	2017-04-11 02:08:54 -0500
+++ ./src/uqm/trans.c	2017-04-11 02:08:43 -0500
@@ -20,7 +20,7 @@
 #include "libs/compiler.h"
 
 
-SIZE sinetab[] =
+SDWORD sinetab[] =
 {
 	-FLT_ADJUST (1.000000),
 	-FLT_ADJUST (0.995185),
diff -ruNp ./src.orig/uqm/units.h ./src/uqm/units.h
--- ./src.orig/uqm/units.h	2017-04-11 02:08:54 -0500
+++ ./src/uqm/units.h	2017-04-11 02:08:43 -0500
@@ -77,9 +77,10 @@ extern int ScreenHeight;
 #define SCALED_ONE (1 << ONE_SHIFT)
 #define DISPLAY_TO_WORLD(x) ((x)<<ONE_SHIFT)
 #define WORLD_TO_DISPLAY(x) ((x)>>ONE_SHIFT)
-#define DISPLAY_ALIGN(x) ((COORD)(x)&~(SCALED_ONE-1))
-#define DISPLAY_ALIGN_X(x) ((COORD)((COUNT)(x)%LOG_SPACE_WIDTH)&~(SCALED_ONE-1))
-#define DISPLAY_ALIGN_Y(y) ((COORD)((COUNT)(y)%LOG_SPACE_HEIGHT)&~(SCALED_ONE-1))
+// JMS_GFX: Changed from COORD to SDWORD and from COUNT to DWORD
+#define DISPLAY_ALIGN(x) ((SDWORD)(x)&~(SCALED_ONE-1))
+#define DISPLAY_ALIGN_X(x) ((SDWORD)((DWORD)(x)%LOG_SPACE_WIDTH)&~(SCALED_ONE-1))
+#define DISPLAY_ALIGN_Y(y) ((SDWORD)((DWORD)(y)%LOG_SPACE_HEIGHT)&~(SCALED_ONE-1))
 
 #define LOG_SPACE_WIDTH \
 		(DISPLAY_TO_WORLD (SPACE_WIDTH) << MAX_REDUCTION)
@@ -186,8 +187,8 @@ universeToLogy (COORD uy)
 										>>(CIRCLE_SHIFT-FACING_SHIFT))
 #define FACING_TO_ANGLE(f) ((f)<<(CIRCLE_SHIFT-FACING_SHIFT))
 
-#define NORMALIZE_ANGLE(a) ((COUNT)((a)&(FULL_CIRCLE-1)))
-#define NORMALIZE_FACING(f) ((COUNT)((f)&((1 << FACING_SHIFT)-1)))
+#define NORMALIZE_ANGLE(a) ((DWORD)((a)&(FULL_CIRCLE-1)))
+#define NORMALIZE_FACING(f) ((DWORD)((f)&((1 << FACING_SHIFT)-1)))
 
 #define DEGREES_TO_ANGLE(d) NORMALIZE_ANGLE((((d) % 360) * FULL_CIRCLE \
 				+ HALF_CIRCLE) / 360)
@@ -200,7 +201,7 @@ universeToLogy (COORD uy)
 #define UNADJUST(x) (SIZE)((x)>>SIN_SHIFT)
 #define ROUND(x,y) ((x)+((x)>=0?((y)>>1):-((y)>>1)))
 
-extern SIZE sinetab[];
+extern SDWORD sinetab[];
 #define SINVAL(a) sinetab[NORMALIZE_ANGLE(a)]
 #define COSVAL(a) SINVAL((a)+QUADRANT)
 #define SINE(a,m) ((SIZE)((((long)SINVAL(a))*(long)(m))>>SIN_SHIFT))
diff -ruNp ./src.orig/uqm/uqmdebug.c ./src/uqm/uqmdebug.c
--- ./src.orig/uqm/uqmdebug.c	2017-04-11 02:08:54 -0500
+++ ./src/uqm/uqmdebug.c	2017-04-11 02:08:43 -0500
@@ -14,8 +14,6 @@
  *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
  */
 
-#if defined(DEBUG) || defined(USE_DEBUG_KEY)
-
 #include "uqmdebug.h"
 
 #include "build.h"
@@ -40,6 +38,107 @@
 #include <stdio.h>
 #include <errno.h>
 
+void (* volatile debugHook) (void) = NULL;
+
+// Move the Flagship to the destination of the autopilot.
+// Should only be called from HyperSpace/QuasiSpace.
+// It can be called from debugHook directly after entering HS/QS though.
+void
+doInstantMove (void)
+{
+	// Move to the new location:
+	if ((GLOBAL (autopilot)).x == ~0 || (GLOBAL (autopilot)).y == ~0)
+	{
+		// If no destination has been selected, use the current location
+		// as the destination.
+		(GLOBAL (autopilot)).x = LOGX_TO_UNIVERSE(GLOBAL_SIS (log_x));
+		(GLOBAL (autopilot)).y = LOGY_TO_UNIVERSE(GLOBAL_SIS (log_y));
+	}
+	else
+	{
+		// A new destination has been selected.
+		GLOBAL_SIS (log_x) = UNIVERSE_TO_LOGX((GLOBAL (autopilot)).x);
+		GLOBAL_SIS (log_y) = UNIVERSE_TO_LOGY((GLOBAL (autopilot)).y);
+	}
+
+	// Check for a solar systems at the destination.
+	if (GET_GAME_STATE (ARILOU_SPACE_SIDE) <= 1)
+	{
+		// If there's a solar system at the destination, enter it.
+		CurStarDescPtr = FindStar (0, &(GLOBAL (autopilot)), 0, 0);
+		if (CurStarDescPtr)
+		{
+			// Leave HyperSpace/QuasiSpace if we're there:
+			SET_GAME_STATE (USED_BROADCASTER, 0);
+			GLOBAL (CurrentActivity) &= ~IN_BATTLE;
+
+			// Enter IP:
+			GLOBAL (ShipFacing) = 0;
+			GLOBAL (ip_planet) = 0;
+			GLOBAL (in_orbit) = 0;
+					// This causes the ship position in IP to be reset.
+			GLOBAL (CurrentActivity) |= START_INTERPLANETARY;
+		}
+	}
+
+	// Turn off the autopilot:
+	(GLOBAL (autopilot)).x = ~0;
+	(GLOBAL (autopilot)).y = ~0;
+}
+
+// playerNr should be 0 or 1
+STARSHIP*
+findPlayerShip (SIZE playerNr)
+{
+	HELEMENT hElement, hNextElement;
+
+	for (hElement = GetHeadElement (); hElement; hElement = hNextElement)
+	{
+		ELEMENT *ElementPtr;
+
+		LockElement (hElement, &ElementPtr);
+		hNextElement = GetSuccElement (ElementPtr);
+					
+		if ((ElementPtr->state_flags & PLAYER_SHIP)	&&
+				ElementPtr->playerNr == playerNr)
+		{
+			STARSHIP *StarShipPtr;
+			GetElementStarShip (ElementPtr, &StarShipPtr);
+			UnlockElement (hElement);
+			return StarShipPtr;
+		}
+		
+		UnlockElement (hElement);
+	}
+	return NULL;
+}
+
+////////////////////////////////////////////////////////////////////////////
+
+void
+resetEnergyBattle (void)
+{
+	STARSHIP *StarShipPtr;
+	COUNT delta;
+	CONTEXT OldContext;
+	
+	if (!(GLOBAL (CurrentActivity) & IN_BATTLE) ||
+			(LOBYTE (GLOBAL (CurrentActivity)) == IN_HYPERSPACE))
+		return;
+	
+	StarShipPtr = findPlayerShip (RPG_PLAYER_NUM);
+	if (StarShipPtr == NULL || StarShipPtr->RaceDescPtr == NULL)
+		return;
+
+	delta = StarShipPtr->RaceDescPtr->ship_info.max_energy -
+			StarShipPtr->RaceDescPtr->ship_info.energy_level;
+
+	OldContext = SetContext (StatusContext);
+	DeltaEnergy (StarShipPtr->hShip, delta);
+	SetContext (OldContext);
+}
+
+#if defined(DEBUG) || defined(USE_DEBUG_KEY)
 
 static void dumpEventCallback (const EVENT *eventPtr, void *arg);
 
@@ -71,7 +170,6 @@ static void dumpPlanetTypeCallback (int
 
 BOOLEAN instantMove = FALSE;
 BOOLEAN disableInteractivity = FALSE;
-void (* volatile debugHook) (void) = NULL;
 void (* volatile doInputDebugHook) (void) = NULL;
 
 
@@ -459,52 +557,6 @@ findFlagshipElement (void)
 }
 #endif
 
-// Move the Flagship to the destination of the autopilot.
-// Should only be called from HyperSpace/QuasiSpace.
-// It can be called from debugHook directly after entering HS/QS though.
-void
-doInstantMove (void)
-{
-	// Move to the new location:
-	if ((GLOBAL (autopilot)).x == ~0 || (GLOBAL (autopilot)).y == ~0)
-	{
-		// If no destination has been selected, use the current location
-		// as the destination.
-		(GLOBAL (autopilot)).x = LOGX_TO_UNIVERSE(GLOBAL_SIS (log_x));
-		(GLOBAL (autopilot)).y = LOGY_TO_UNIVERSE(GLOBAL_SIS (log_y));
-	}
-	else
-	{
-		// A new destination has been selected.
-		GLOBAL_SIS (log_x) = UNIVERSE_TO_LOGX((GLOBAL (autopilot)).x);
-		GLOBAL_SIS (log_y) = UNIVERSE_TO_LOGY((GLOBAL (autopilot)).y);
-	}
-
-	// Check for a solar systems at the destination.
-	if (GET_GAME_STATE (ARILOU_SPACE_SIDE) <= 1)
-	{
-		// If there's a solar system at the destination, enter it.
-		CurStarDescPtr = FindStar (0, &(GLOBAL (autopilot)), 0, 0);
-		if (CurStarDescPtr)
-		{
-			// Leave HyperSpace/QuasiSpace if we're there:
-			SET_GAME_STATE (USED_BROADCASTER, 0);
-			GLOBAL (CurrentActivity) &= ~IN_BATTLE;
-
-			// Enter IP:
-			GLOBAL (ShipFacing) = 0;
-			GLOBAL (ip_planet) = 0;
-			GLOBAL (in_orbit) = 0;
-					// This causes the ship position in IP to be reset.
-			GLOBAL (CurrentActivity) |= START_INTERPLANETARY;
-		}
-	}
-
-	// Turn off the autopilot:
-	(GLOBAL (autopilot)).x = ~0;
-	(GLOBAL (autopilot)).y = ~0;
-}
-
 ////////////////////////////////////////////////////////////////////////////
 
 void
@@ -1467,35 +1519,6 @@ depositQualityString (BYTE quality)
 
 ////////////////////////////////////////////////////////////////////////////
 
-// playerNr should be 0 or 1
-STARSHIP*
-findPlayerShip (SIZE playerNr)
-{
-	HELEMENT hElement, hNextElement;
-
-	for (hElement = GetHeadElement (); hElement; hElement = hNextElement)
-	{
-		ELEMENT *ElementPtr;
-
-		LockElement (hElement, &ElementPtr);
-		hNextElement = GetSuccElement (ElementPtr);
-					
-		if ((ElementPtr->state_flags & PLAYER_SHIP)	&&
-				ElementPtr->playerNr == playerNr)
-		{
-			STARSHIP *StarShipPtr;
-			GetElementStarShip (ElementPtr, &StarShipPtr);
-			UnlockElement (hElement);
-			return StarShipPtr;
-		}
-		
-		UnlockElement (hElement);
-	}
-	return NULL;
-}
-
-////////////////////////////////////////////////////////////////////////////
-
 void
 resetCrewBattle (void)
 {
@@ -1519,29 +1542,6 @@ resetCrewBattle (void)
 	SetContext (OldContext);
 }
 
-void
-resetEnergyBattle (void)
-{
-	STARSHIP *StarShipPtr;
-	COUNT delta;
-	CONTEXT OldContext;
-	
-	if (!(GLOBAL (CurrentActivity) & IN_BATTLE) ||
-			(LOBYTE (GLOBAL (CurrentActivity)) == IN_HYPERSPACE))
-		return;
-	
-	StarShipPtr = findPlayerShip (RPG_PLAYER_NUM);
-	if (StarShipPtr == NULL || StarShipPtr->RaceDescPtr == NULL)
-		return;
-
-	delta = StarShipPtr->RaceDescPtr->ship_info.max_energy -
-			StarShipPtr->RaceDescPtr->ship_info.energy_level;
-
-	OldContext = SetContext (StatusContext);
-	DeltaEnergy (StarShipPtr->hShip, delta);
-	SetContext (OldContext);
-}
-
 ////////////////////////////////////////////////////////////////////////////
 
 // This function should help in making sure that gamestr.h matches
diff -ruNp ./src.orig/uqm/uqmdebug.h ./src/uqm/uqmdebug.h
--- ./src.orig/uqm/uqmdebug.h	2017-04-11 02:08:54 -0500
+++ ./src/uqm/uqmdebug.h	2017-04-11 02:08:43 -0500
@@ -14,8 +14,6 @@
  *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
  */
 
-#if !defined(_DEBUG_H) && (defined(DEBUG) || defined(USE_DEBUG_KEY))
-#define _DEBUG_H
 
 #include "clock.h"
 #include "planets/planets.h"
@@ -24,16 +22,29 @@
 
 #include <stdio.h>
 
+// If a function is assigned to this, it will be called from the
+// Starcon2Main thread, in the main game loop.
+extern void (* volatile debugHook) (void);
+
+// Move the Flagship to the destination of the autopilot.
+// Should only be called from HS/QS.
+// It can be called from debugHook directly after entering HS/QS though.
+void doInstantMove (void);
+
+// Find a player ship. Setting playerNr to non-0 is only meaningful in battle.
+STARSHIP* findPlayerShip (SIZE playerNr);
+
+// Resets the energy of the first player (the bottom one) to its maximum.
+void resetEnergyBattle(void);
+
+#if !defined(_DEBUG_H) && (defined(DEBUG) || defined(USE_DEBUG_KEY))
+#define _DEBUG_H
 
 // If set to true, interactive routines that are called (indirectly) in debug
 // functions are a no-op.
 extern BOOLEAN disableInteractivity;
 
 // If a function is assigned to this, it will be called from the
-// Starcon2Main thread, in the main game loop.
-extern void (* volatile debugHook) (void);
-
-// If a function is assigned to this, it will be called from the
 // Starcon2Main thread, in doInput().
 extern void (* volatile doInputDebugHook) (void);
 
@@ -66,12 +77,6 @@ void showSpheres (void);
 // Make the ships of all races available for building at the shipyard.
 void activateAllShips (void);
 
-// Move the Flagship to the destination of the autopilot.
-// Should only be called from HS/QS.
-// It can be called from debugHook directly after entering HS/QS though.
-void doInstantMove (void);
-
-
 // Call a function for all stars.
 void forAllStars (void (*callback) (STAR_DESC *, void *), void *arg);
 // Call a function for all planets in a star system.
@@ -166,17 +171,9 @@ const char *densityString (BYTE density)
 // Get a string describing the quality of a deposit.
 const char *depositQualityString (BYTE quality);
 
-
-// Find a player ship. Setting playerNr to non-0 is only meaningful in battle.
-STARSHIP* findPlayerShip (SIZE playerNr);
-
 // Resets the crew of the first player (the bottom one) to its maximum.
 void resetCrewBattle(void);
 
-// Resets the energy of the first player (the bottom one) to its maximum.
-void resetEnergyBattle(void);
-
-
 // Move instantly across hyperspace/quasispace.
 extern BOOLEAN instantMove;
 
diff -ruNp ./src.orig/uqm.c ./src/uqm.c
--- ./src.orig/uqm.c	2017-04-11 02:08:54 -0500
+++ ./src/uqm.c	2017-04-11 02:08:43 -0500
@@ -127,6 +127,18 @@ struct options_struct
 	DECL_CONFIG_OPTION(float, sfxVolumeScale);
 	DECL_CONFIG_OPTION(float, speechVolumeScale);
 	DECL_CONFIG_OPTION(bool, safeMode);
+ 	DECL_CONFIG_OPTION(bool, cheatMode); // JMS
+	DECL_CONFIG_OPTION(bool, godMode); // Serosis
+	DECL_CONFIG_OPTION(bool, timeDilation); // Serosis
+	DECL_CONFIG_OPTION(bool, bubbleWarp); // Serosis
+	DECL_CONFIG_OPTION(bool, roseBud); // Serosis
+	DECL_CONFIG_OPTION(bool, unlockShips); // Serosis
+	DECL_CONFIG_OPTION(bool, headStart); // Serosis
+	DECL_CONFIG_OPTION(bool, unlockUpgrades);
+	DECL_CONFIG_OPTION(bool, landerMods);
+	DECL_CONFIG_OPTION(bool, fastForward);
+	DECL_CONFIG_OPTION(bool, skipIntro);
+	DECL_CONFIG_OPTION(bool, FMV);
 
 #define INIT_CONFIG_OPTION(name, val) \
 	{ val, false }
@@ -235,10 +247,10 @@ main (int argc, char *argv[])
 		/* .addons = */             NULL,
 		/* .numAddons = */          0,
 
-		INIT_CONFIG_OPTION(  opengl,            false ),
+		INIT_CONFIG_OPTION(  opengl,            true ),
 		INIT_CONFIG_OPTION2( resolution,        640, 480 ),
 		INIT_CONFIG_OPTION(  fullscreen,        false ),
-		INIT_CONFIG_OPTION(  scanlines,         false ),
+		INIT_CONFIG_OPTION(  scanlines,         true ),
 		INIT_CONFIG_OPTION(  scaler,            0 ),
 		INIT_CONFIG_OPTION(  showFps,           false ),
 		INIT_CONFIG_OPTION(  keepAspectRatio,   false ),
@@ -251,7 +263,7 @@ main (int argc, char *argv[])
 		INIT_CONFIG_OPTION(  whichMenu,         OPT_PC ),
 		INIT_CONFIG_OPTION(  whichFonts,        OPT_PC ),
 		INIT_CONFIG_OPTION(  whichIntro,        OPT_PC ),
-		INIT_CONFIG_OPTION(  whichShield,       OPT_PC ),
+		INIT_CONFIG_OPTION(  whichShield,       OPT_3DO ),
 		INIT_CONFIG_OPTION(  smoothScroll,      OPT_PC ),
 		INIT_CONFIG_OPTION(  meleeScale,        TFB_SCALE_TRILINEAR ),
 		INIT_CONFIG_OPTION(  subtitles,         true ),
@@ -260,6 +272,18 @@ main (int argc, char *argv[])
 		INIT_CONFIG_OPTION(  sfxVolumeScale,    1.0f ),
 		INIT_CONFIG_OPTION(  speechVolumeScale, 1.0f ),
 		INIT_CONFIG_OPTION(  safeMode,          false ),
+		INIT_CONFIG_OPTION(  cheatMode,			false ),
+		INIT_CONFIG_OPTION(  godMode,			false ), //Serosis
+		INIT_CONFIG_OPTION(  timeDilation,		false ),
+		INIT_CONFIG_OPTION(  bubbleWarp,		false ),
+		INIT_CONFIG_OPTION(  roseBud,			false ), 
+		INIT_CONFIG_OPTION(  unlockShips,		false ), 
+		INIT_CONFIG_OPTION(  headStart,			false ), 
+		INIT_CONFIG_OPTION(  unlockUpgrades,	false ),
+		INIT_CONFIG_OPTION(  landerMods,		false ),
+		INIT_CONFIG_OPTION(  fastForward,		false ),
+		INIT_CONFIG_OPTION(  skipIntro,			false ),
+		INIT_CONFIG_OPTION(  FMV,				false ),
 	};
 	struct options_struct defaults = options;
 	int optionsResult;
@@ -382,6 +406,18 @@ main (int argc, char *argv[])
 	sfxVolumeScale = options.sfxVolumeScale.value;
 	speechVolumeScale = options.speechVolumeScale.value;
 	optAddons = options.addons;
+ 	optCheatMode = options.cheatMode.value; // JMS
+	optGodMode = options.godMode.value; // Serosis
+	optTimeDilation = options.timeDilation.value;
+	optBubbleWarp = options.bubbleWarp.value;
+	optRoseBud = options.roseBud.value;
+	optUnlockShips = options.unlockShips.value;
+	optHeadStart = options.headStart.value;
+	optUnlockUpgrades = options.unlockUpgrades.value;
+	optLanderMods = options.landerMods.value;
+	optFastForward = options.fastForward.value;
+	optSkipIntro = options.skipIntro.value;
+	optFMV = options.FMV.value;
 
 	prepareContentDir (options.contentDir, options.addonDir, argv[0]);
 	prepareMeleeDir ();
@@ -491,7 +527,7 @@ saveErrorV (const char *fmt, va_list lis
 		++len;
 		--left;
 	}
-	vsnprintf (errBuffer + len, left, fmt, list);
+	uqm_vsnprintf (errBuffer + len, left, fmt, list);
 	errBuffer[sizeof (errBuffer) - 1] = '\0';
 }
 
@@ -639,6 +675,18 @@ getUserConfigOptions (struct options_str
 	getVolumeConfigValue (&options->musicVolumeScale, "config.musicvol");
 	getVolumeConfigValue (&options->sfxVolumeScale, "config.sfxvol");
 	getVolumeConfigValue (&options->speechVolumeScale, "config.speechvol");
+	getBoolConfigValue (&options->cheatMode, "config.cheatMode");
+	getBoolConfigValue (&options->godMode, "config.godMode"); //Serosis
+	getBoolConfigValue (&options->timeDilation, "config.timeDilation");
+	getBoolConfigValue (&options->bubbleWarp, "config.bubbleWarp"); 
+	getBoolConfigValue (&options->roseBud, "config.roseBud"); 
+	getBoolConfigValue (&options->unlockShips, "config.unlockShips"); 
+	getBoolConfigValue (&options->headStart, "config.headStart"); 
+	getBoolConfigValue (&options->unlockUpgrades, "config.unlockUpgrades");
+	getBoolConfigValue (&options->landerMods, "config.landerMods");
+	getBoolConfigValue (&options->fastForward, "config.fastForward");
+	getBoolConfigValue (&options->skipIntro, "config.skipIntro");
+	getBoolConfigValue (&options->FMV, "config.FMV");
 	
 	if (res_IsInteger ("config.player1control"))
 	{
@@ -678,6 +726,18 @@ enum
 	ADDONDIR_OPT,
 	ACCEL_OPT,
 	SAFEMODE_OPT,
+	CHEATMODE_OPT, //Serosis
+	GODMODE_OPT, //Serosis
+	TDM_OPT, //Serosis
+	BWARP_OPT, //Serosis
+	ROSEBUD_OPT, //Serosis
+	UNLOCKSHIPS_OPT, //Serosis
+	HEADSTART_OPT, //Serosis
+	UPGRADES_OPT,
+	LANDERCHT_OPT,
+	FASTFORWARD_OPT,
+	SKIPINTRO_OPT,
+	FMV_OPT,
 #ifdef NETPLAY
 	NETHOST1_OPT,
 	NETPORT1_OPT,
@@ -725,6 +785,18 @@ static struct option longOptions[] =
 	{"addondir", 1, NULL, ADDONDIR_OPT},
 	{"accel", 1, NULL, ACCEL_OPT},
 	{"safe", 0, NULL, SAFEMODE_OPT},
+	{"cheatmode", 0, NULL, CHEATMODE_OPT}, //Serosis
+	{"godmode", 0, NULL, GODMODE_OPT}, //Serosis
+	{"timedilation", 0, NULL, TDM_OPT}, //Serosis
+	{"bubblewarp", 0, NULL, BWARP_OPT}, //Serosis
+	{"rosebud", 0, NULL, ROSEBUD_OPT}, //Serosis
+	{"unlockships", 0, NULL, UNLOCKSHIPS_OPT}, //Serosis
+	{"headstart", 0, NULL, HEADSTART_OPT}, //Serosis
+	{"unlockupgrades", 0, NULL, UPGRADES_OPT},
+	{"landermods", 0, NULL, LANDERCHT_OPT},
+	{"fastforward", 0, NULL, FASTFORWARD_OPT},
+	{"skipintro", 0, NULL, SKIPINTRO_OPT},
+	{"fmv", 0, NULL, FMV_OPT},
 #ifdef NETPLAY
 	{"nethost1", 1, NULL, NETHOST1_OPT},
 	{"netport1", 1, NULL, NETPORT1_OPT},
@@ -1013,9 +1085,45 @@ parseOptions (int argc, char *argv[], st
 				}
 				break;
 			}
-	                case SAFEMODE_OPT:
+			case SAFEMODE_OPT:
 				setBoolOption (&options->safeMode, true);
 				break;
+			case CHEATMODE_OPT:
+				setBoolOption (&options->cheatMode, true); //Serosis
+				break;
+			case GODMODE_OPT:
+				setBoolOption (&options->godMode, true); //Serosis
+				break;
+			case TDM_OPT:
+				setBoolOption (&options->timeDilation, true); //Serosis
+				break;
+			case BWARP_OPT:
+				setBoolOption (&options->bubbleWarp, true); //Serosis
+				break;
+			case ROSEBUD_OPT:
+				setBoolOption (&options->roseBud, true); //Serosis
+				break;
+			case UNLOCKSHIPS_OPT:
+				setBoolOption (&options->unlockShips, true); //Serosis
+				break;
+			case HEADSTART_OPT:
+				setBoolOption (&options->headStart, true); //Serosis
+ 				break;
+			case UPGRADES_OPT:
+				setBoolOption (&options->unlockUpgrades, true);
+				break;
+			case LANDERCHT_OPT:
+				setBoolOption (&options->landerMods, true);
+				break;
+			case FASTFORWARD_OPT:
+				setBoolOption (&options->fastForward, true);
+				break;
+			case SKIPINTRO_OPT:
+				setBoolOption (&options->skipIntro, true);
+				break;
+			case FMV_OPT:
+				setBoolOption (&options->FMV, true);
+				break;
 #ifdef NETPLAY
 			case NETHOST1_OPT:
 				netplayOptions.peer[0].isServer = false;
@@ -1204,6 +1312,40 @@ usage (FILE *out, const struct options_s
 	log_add (log_User, "  --scroll    : ff/frev during comm.  pc=per-page, "
 			"3do=smooth (default %s)",
 			choiceOptString (&defaults->smoothScroll));
+	log_add (log_User, "The following options are for the Mega Mod"); // Serosis
+	log_add (log_User, "  --cheatmode : Stops Kohr-Ah advancing.    (default %s)",
+			boolOptString (&defaults->cheatMode));
+	log_add (log_User, "  --godmode : Player ships and lander invulnerable. "
+			"Also refills energy every shot during melee.    (default %s)",
+			boolOptString (&defaults->godMode));
+	log_add (log_User, "  --timedilation : Increases time by a factor of 6. "
+			"IP=3 Minutes HS=30 Seconds    (default %s)",
+			boolOptString (&defaults->timeDilation));
+	log_add (log_User, "  --bubblewarp : Instantaneous travel to any point on "
+			"the Starmap.    (default %s)",
+			boolOptString (&defaults->bubbleWarp));
+	log_add (log_User, "  --rosebud : Get 1,000 R.U. and Credits when you ask the "
+			"Melnorme why their bridge is purple.    (default %s)",
+			boolOptString (&defaults->roseBud));
+	log_add (log_User, "  --unlockships : Allows you to purchase ships that you can't "
+			"normally acquire in the main game.    (default %s)",
+			boolOptString (&defaults->unlockShips));
+	log_add (log_User, "  --headstart : Equips your ship with full thrusters and jets, "
+			"two ion bolt guns, two cargo bays, two crew pods, four dynamos, two full fuel tanks, "
+			"max landers, 1000 radioactives, and 1000 bio-units.    (default %s)",
+			boolOptString (&defaults->headStart));
+	log_add (log_User, "  --unlockupgrades : Unlocks every upgrade for your flagship "
+			"and landers.    (default %s)",
+			boolOptString (&defaults->unlockUpgrades));
+	log_add (log_User, "  --landermods : Makes your landers have pin-point accuracy "
+			"when landing and doubles storage capacity.   (default %s)",
+			boolOptString (&defaults->landerMods));
+	log_add (log_User, "  --fastforward : Speeds up time by a factor of 5    (default %s)",
+			boolOptString (&defaults->fastForward));
+	log_add (log_User, "  --skipintro : Skips the intro    (default %s)",
+			boolOptString (&defaults->skipIntro));
+	log_add (log_User, "  --fmv : Adds Logo and Commercial 3DO videos    (default %s)",
+			boolOptString (&defaults->FMV));
 	log_setOutput (old);
 }
 
diff -ruNp ./src.orig/uqmversion.h ./src/uqmversion.h
--- ./src.orig/uqmversion.h	2017-04-11 02:08:54 -0500
+++ ./src/uqmversion.h	2017-04-11 02:08:43 -0500
@@ -21,9 +21,9 @@
 #define UQM_MAJOR_VERSION_S  "0"
 #define UQM_MINOR_VERSION     7
 #define UQM_MINOR_VERSION_S  "7"
-#define UQM_PATCH_VERSION     0
-#define UQM_PATCH_VERSION_S  "0"
-#define UQM_EXTRA_VERSION    ""
+#define UQM_PATCH_VERSION     2
+#define UQM_PATCH_VERSION_S  "2"
+#define UQM_EXTRA_VERSION    "e MegaMod"
 /* The final version is interpreted as:
  * printf ("%d.%d.%d%s", UQM_MAJOR_VERSION, UQM_MINOR_VERSION,
  * 		UQM_PATCH_VERSION, UQM_EXTRA_VERSION);
