diff -ruNp ./src.vc10/libs/uio/utils.c ./src/libs/uio/utils.c
--- ./src.vc10/libs/uio/utils.c	2017-03-05 21:37:19 -0600
+++ ./src/libs/uio/utils.c	2017-02-25 17:11:19 -0600
@@ -391,7 +391,7 @@ snprintf(char *str, size_t size, const c
 // MSVC does not have vsnprintf(). It does have a _vsnprintf(), but it does
 // not \0-terminate a truncated string as the C standard prescribes.
 static inline int
-uqm_vsnprintf(char *str, size_t size, const char *format, va_list args)
+vsnprintf(char *str, size_t size, const char *format, va_list args)
 {
 	int result = _vsnprintf (str, size, format, args);
 	if (str != NULL && size != 0)
@@ -425,7 +425,7 @@ uio_vasprintf(const char *format, va_lis
 	}
 
 	for (;;) {
-		int printResult = uqm_vsnprintf(buf, bufSize, format, args);
+		int printResult = vsnprintf(buf, bufSize, format, args);
 		if (printResult < 0) {
 			// This means the buffer was not large enough, but vsnprintf()
 			// does not give us any clue on how large it should be.
diff -ruNp ./src.vc10/port.c ./src/port.c
--- ./src.vc10/port.c	2017-03-05 21:37:19 -0600
+++ ./src/port.c	2017-02-25 17:11:19 -0600
@@ -115,8 +115,8 @@ readdir_r(DIR *dirp, struct dirent *entr
 #endif
 
 #ifdef _MSC_VER
-// MSVC does not have snprintf() and uqm_vsnprintf(). It does have a _snprintf()
-// and _uqm_vsnprintf(), but these do not terminate a truncated string as
+// MSVC does not have snprintf() and vsnprintf(). It does have a _snprintf()
+// and _vsnprintf(), but these do not terminate a truncated string as
 // the C standard prescribes.
 int
 snprintf(char *str, size_t size, const char *format, ...)
@@ -134,7 +134,7 @@ snprintf(char *str, size_t size, const c
 }
 
 int
-uqm_vsnprintf(char *str, size_t size, const char *format, va_list args)
+vsnprintf(char *str, size_t size, const char *format, va_list args)
 {
 	int result = _vsnprintf (str, size, format, args);
 	if (str != NULL && size != 0)
diff -ruNp ./src.vc10/port.h ./src/port.h
--- ./src.vc10/port.h	2017-03-05 21:37:19 -0600
+++ ./src/port.h	2017-02-25 17:11:19 -0600
@@ -162,7 +162,7 @@ typedef unsigned short mode_t;
 #	include <stdarg.h>
 // Defined in port.c
 int snprintf(char *str, size_t size, const char *format, ...);
-int uqm_vsnprintf(char *str, size_t size, const char *format, va_list args);
+int vsnprintf(char *str, size_t size, const char *format, va_list args);
 #endif  /* _MSC_VER */
 
 // setenv()
diff -ruNp ./src.vc10/uqm.c ./src/uqm.c
--- ./src.vc10/uqm.c	2017-03-05 21:37:24 -0600
+++ ./src/uqm.c	2017-03-05 21:43:04 -0600
@@ -491,7 +521,7 @@ saveErrorV (const char *fmt, va_list lis
 		++len;
 		--left;
 	}
-	uqm_vsnprintf (errBuffer + len, left, fmt, list);
+	vsnprintf (errBuffer + len, left, fmt, list);
 	errBuffer[sizeof (errBuffer) - 1] = '\0';
 }
 
